<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en-US" xml:lang="en-US">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Dylan G. E. Gomes" />
  <meta name="author" content="Cole B. Brookson" />
  <meta name="author" content="Robert Crystal-Ornelas" />
  <meta name="author" content="Ali Guncan" />
  <meta name="author" content="Brandon P.M. Edwards" />
  <meta name="author" content="Kaitlyn M. Gaynor" />
  <meta name="author" content="Vivienne Foroughirad" />
  <meta name="author" content="Katherine Hébert" />
  <meta name="author" content="Emma J. Hudgins" />
  <meta name="author" content="Saeed Shafiei Sabet" />
  <meta name="author" content="Eric R. Scott" />
  <meta name="author" content="Allison D. Binley" />
  <meta name="dcterms.date" content="2022-04-12" />
  <meta name="keywords" content="Github, Reproducible research, Collaboration, Ecology and evolution" />
  <title>Not just for programmers: A friendly guide on the versatility/benefits of GitHub for accelerating collaborative research in Ecology and Evolution</title>
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
    }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Not just for programmers: A friendly guide on the versatility/benefits of GitHub for accelerating collaborative research in Ecology and Evolution" />
  <meta name="citation_title" content="Not just for programmers: A friendly guide on the versatility/benefits of GitHub for accelerating collaborative research in Ecology and Evolution" />
  <meta property="og:title" content="Not just for programmers: A friendly guide on the versatility/benefits of GitHub for accelerating collaborative research in Ecology and Evolution" />
  <meta property="twitter:title" content="Not just for programmers: A friendly guide on the versatility/benefits of GitHub for accelerating collaborative research in Ecology and Evolution" />
  <meta name="dc.date" content="2022-04-12" />
  <meta name="citation_publication_date" content="2022-04-12" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Dylan G. E. Gomes" />
  <meta name="citation_author_institution" content="Cooperative Institute for Marine Resources Studies, Hatfield Marine Science Center, Oregon State University, Newport, OR, United States" />
  <meta name="citation_author_orcid" content="0000-0002-2642-3728" />
  <meta name="citation_author" content="Cole B. Brookson" />
  <meta name="citation_author_institution" content="Department of Biological Sciences, University of Alberta, Edmonton, AB, Canada" />
  <meta name="citation_author_orcid" content="0000-0003-1237-4096" />
  <meta name="citation_author" content="Robert Crystal-Ornelas" />
  <meta name="citation_author_institution" content="Earth and Environmental Sciences Area, Lawrence Berkeley National Laboratory, Berkeley, CA 94720, USA" />
  <meta name="citation_author_orcid" content="0000-0002-6339-1139" />
  <meta name="twitter:creator" content="@rob_c_ornelas" />
  <meta name="citation_author" content="Ali Guncan" />
  <meta name="citation_author_institution" content="Department of Plant Protection, Faculty of Agriculture, Ordu University, 52200, Ordu, Turkey" />
  <meta name="citation_author_orcid" content="0000-0003-1765-648X" />
  <meta name="twitter:creator" content="@aliguncan" />
  <meta name="citation_author" content="Brandon P.M. Edwards" />
  <meta name="citation_author_institution" content="Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada" />
  <meta name="citation_author_orcid" content="0000-0003-0865-3076" />
  <meta name="citation_author" content="Kaitlyn M. Gaynor" />
  <meta name="citation_author_institution" content="Departments of Zoology and Botany, University of British Columbia, Vancouver, BC, Canada" />
  <meta name="citation_author_institution" content="National Center for Ecological Analysis and Synthesis, Santa Barbara, CA 93101, USA" />
  <meta name="citation_author_orcid" content="0000-0002-5747-0543" />
  <meta name="twitter:creator" content="@kaitlyngaynor" />
  <meta name="citation_author" content="Vivienne Foroughirad" />
  <meta name="citation_author_institution" content="Department of Biology, Georgetown University, Washington, DC, USA" />
  <meta name="citation_author_orcid" content="0000-0002-8656-7440" />
  <meta name="twitter:creator" content="@vforoughirad" />
  <meta name="citation_author" content="Katherine Hébert" />
  <meta name="citation_author_institution" content="Département de biologie, Université de Sherbrooke, Québec, Canada" />
  <meta name="citation_author_orcid" content="0000-0001-7866-6775" />
  <meta name="twitter:creator" content="@hebert_kat" />
  <meta name="citation_author" content="Emma J. Hudgins" />
  <meta name="citation_author_institution" content="Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada" />
  <meta name="citation_author_orcid" content="0000-0002-8402-5111" />
  <meta name="citation_author" content="Saeed Shafiei Sabet" />
  <meta name="citation_author_institution" content="Fisheries Department, Faculty of Natural Resources, University of Guilan, Sowmeh Sara, Iran" />
  <meta name="citation_author_orcid" content="0000-0001-5919-2527" />
  <meta name="twitter:creator" content="@SaeedSHSABET" />
  <meta name="citation_author" content="Eric R. Scott" />
  <meta name="citation_author_institution" content="Department of Wildlife Ecology and Conservation, University of Florida, Gainesville, FL, USA" />
  <meta name="citation_author_orcid" content="0000-0002-7430-7879" />
  <meta name="citation_author" content="Allison D. Binley" />
  <meta name="citation_author_institution" content="Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada" />
  <meta name="citation_author_orcid" content="0000-0001-8790-9935" />
  <meta name="twitter:creator" content="@AllisonBinley" />
  <link rel="canonical" href="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta property="og:url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta property="twitter:url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta name="citation_fulltext_html_url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/" />
  <meta name="citation_pdf_url" content="https://SORTEE-Github-Hackathon.github.io/manuscript/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://SORTEE-Github-Hackathon.github.io/manuscript/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://SORTEE-Github-Hackathon.github.io/manuscript/v/4878dcbf88f1bc507dbc1b4d524ef9b848cac0f9/" />
  <meta name="manubot_html_url_versioned" content="https://SORTEE-Github-Hackathon.github.io/manuscript/v/4878dcbf88f1bc507dbc1b4d524ef9b848cac0f9/" />
  <meta name="manubot_pdf_url_versioned" content="https://SORTEE-Github-Hackathon.github.io/manuscript/v/4878dcbf88f1bc507dbc1b4d524ef9b848cac0f9/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<header id="title-block-header">
<h1 class="title">Not just for programmers: A friendly guide on the versatility/benefits of GitHub for accelerating collaborative research in Ecology and Evolution</h1>
</header>
<p><small><em>
This manuscript
(<a href="https://SORTEE-Github-Hackathon.github.io/manuscript/v/4878dcbf88f1bc507dbc1b4d524ef9b848cac0f9/">permalink</a>)
was automatically generated
from <a href="https://github.com/SORTEE-Github-Hackathon/manuscript/tree/4878dcbf88f1bc507dbc1b4d524ef9b848cac0f9">SORTEE-Github-Hackathon/manuscript@4878dcb</a>
on April 12, 2022.
</em></small></p>
<h2 id="authors">Authors</h2>
<ul>
<li><p><strong>Dylan G. E. Gomes</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-2642-3728">0000-0002-2642-3728</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/dylangomes">dylangomes</a><br>
<small>
Cooperative Institute for Marine Resources Studies, Hatfield Marine Science Center, Oregon State University, Newport, OR, United States
</small></p></li>
<li><p><strong>Cole B. Brookson</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1237-4096">0000-0003-1237-4096</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/colebrookson">colebrookson</a><br>
<small>
Department of Biological Sciences, University of Alberta, Edmonton, AB, Canada
</small></p></li>
<li><p><strong>Robert Crystal-Ornelas</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-6339-1139">0000-0002-6339-1139</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/robcrystalornelas">robcrystalornelas</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/rob_c_ornelas">rob_c_ornelas</a><br>
<small>
Earth and Environmental Sciences Area, Lawrence Berkeley National Laboratory, Berkeley, CA 94720, USA
</small></p></li>
<li><p><strong>Ali Guncan</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-1765-648X">0000-0003-1765-648X</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Aguncan">Aguncan</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/aliguncan">aliguncan</a><br>
<small>
Department of Plant Protection, Faculty of Agriculture, Ordu University, 52200, Ordu, Turkey
</small></p></li>
<li><p><strong>Brandon P.M. Edwards</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0003-0865-3076">0000-0003-0865-3076</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/BrandonEdwards">BrandonEdwards</a><br>
<small>
Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada
</small></p></li>
<li><p><strong>Kaitlyn M. Gaynor</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-5747-0543">0000-0002-5747-0543</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/kaitlyngaynor">kaitlyngaynor</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/kaitlyngaynor">kaitlyngaynor</a><br>
<small>
Departments of Zoology and Botany, University of British Columbia, Vancouver, BC, Canada; National Center for Ecological Analysis and Synthesis, Santa Barbara, CA 93101, USA
</small></p></li>
<li><p><strong>Vivienne Foroughirad</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8656-7440">0000-0002-8656-7440</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/vjf2">vjf2</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/vforoughirad">vforoughirad</a><br>
<small>
Department of Biology, Georgetown University, Washington, DC, USA
</small></p></li>
<li><p><strong>Katherine Hébert</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-7866-6775">0000-0001-7866-6775</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/katherinehebert">katherinehebert</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/hebert_kat">hebert_kat</a><br>
<small>
Département de biologie, Université de Sherbrooke, Québec, Canada
</small></p></li>
<li><p><strong>Emma J. Hudgins</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-8402-5111">0000-0002-8402-5111</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/emmajhudgins">emmajhudgins</a><br>
<small>
Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada
</small></p></li>
<li><p><strong>Saeed Shafiei Sabet</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-5919-2527">0000-0001-5919-2527</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/shafieisabets">shafieisabets</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/SaeedSHSABET">SaeedSHSABET</a><br>
<small>
Fisheries Department, Faculty of Natural Resources, University of Guilan, Sowmeh Sara, Iran
</small></p></li>
<li><p><strong>Eric R. Scott</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0002-7430-7879">0000-0002-7430-7879</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/Aariq">Aariq</a><br>
<small>
Department of Wildlife Ecology and Conservation, University of Florida, Gainesville, FL, USA
</small></p></li>
<li><p><strong>Allison D. Binley</strong><br>
<img src="images/orcid.svg" class="inline_icon" width="16" height="16" alt="ORCID icon" />
<a href="https://orcid.org/0000-0001-8790-9935">0000-0001-8790-9935</a>
· <img src="images/github.svg" class="inline_icon" width="16" height="16" alt="GitHub icon" />
<a href="https://github.com/adbinley">adbinley</a>
· <img src="images/twitter.svg" class="inline_icon" width="16" height="16" alt="Twitter icon" />
<a href="https://twitter.com/AllisonBinley">AllisonBinley</a><br>
<small>
Department of Biology, Carleton University, Ottawa, ON K1S 5B6, Canada
</small></p></li>
</ul>
<h2 class="page_break_before" id="abstract">Abstract</h2>
<p>Ecological and evolutionary researchers have become critically dependent on computional code to achieve scientific elaboration.
With this, the use of efficient forms to collaborate, share and reproduce code has become fundamental.
GitHub is an online, cloud-based service that can help researchers to track, organize, discuss, share and collaborate on software and code.
While scientists in natural sciences have been encouraged to use GitHub, its adoption is not widespread due to the lack of domain-specific information and guidelines.
To help ecology and evolutionary researchers adopt useful features from Github in their own workflows, we review thirteen practical ways to use the platform.
We outline features ranging from low to high technical difficulty: storing code, managing projects, coding collaboratively, conducting peer review, and writing a manuscript.
Because not all GitHub features are essentially useful for every researcher, we allocate existing development and collaborative workflows within Github – such as, tracking tasks, discussing edits to a manuscript, and submitting code to a repository – to the most common research development roles in ecology and evolution (project manager, coauthor, code contributor).
As more ecologists and evolutionary biologists migrate their workflows to GitHub, the broader scientific community stands to benefit from disciplines that push the boundaries of collaborative, transparent, and open research.</p>
<h2 class="page_break_before" id="introduction">Introduction</h2>
<!-- ### General introduction to the use of collaborative computational resources by scientists -->
<!-- *Contributors to this section: PHPB* -->
<p>Most scientists, including ecologists and evolutionary biologists, have become critically dependent on computational tools in their research <span class="citation" data-cites="fJWFe93e">[see <a href="#ref-fJWFe93e" role="doc-biblioref">1</a>]</span>.
Researchers now write and use code as part of their scientific workflow to perform a wide-variety of tasks ranging from data management, data analysis, study replication, to the application and the development of tools for hypothesis testing.
This code-dependent workflow imposes steep requirements towards an efficient, well-documented process in the publication and collaboration of maintainable scientific code <span class="citation" data-cites="1Kqna6l2">[<a href="#ref-1Kqna6l2" role="doc-biblioref">2</a>]</span>.
To facilitate this process, scientists have been increasingly borrowing and adopting tools from information system technology, such as cloud-based services for documentation and version control [<em>e.g.</em> from the Google Suite (with Docs, Sheets and Drive), the Microsoft Suite (with Word, Excel and OneDrive), and Github] <span class="citation" data-cites="doi">[<a href="#ref-doi" role="doc-biblioref"><strong>doi?</strong></a> 10.1038/538127a]</span>.
However, most researchers lack exposure to adequate software development practices and are required to dedicate valuable time and effort to self-teach the use of research-facilitating tools, and thus may find practical barriers when applying adequate standards to maintain their scientific code <span class="citation" data-cites="10SpoByIw O6UbstGG fJWFe93e">[<a href="#ref-fJWFe93e" role="doc-biblioref">1</a>,<a href="#ref-10SpoByIw" role="doc-biblioref">3</a>,<a href="#ref-O6UbstGG" role="doc-biblioref">4</a>]</span>.
Here, we review and discuss one of the most used web-based platforms for computational version control and collaboration, GitHub, and provide researchers in ecology and evolutionary biology (EEB) with practical workflows aimed at facilitating their scientific code and management process.</p>
<!-- ### General background about GitHub and Git -->
<!-- *Contributors to this section: RCO, SSHS, PHPB* -->
<p>With over 73 million registered users, GitHub is the most common web-platform used for collaborating on computer code <span class="citation" data-cites="nwCtHDCn">[<a href="#ref-nwCtHDCn" role="doc-biblioref">5</a>]</span>.
<a href="https://github.com/">GitHub</a> builds on the <a href="https://git-scm.com/">Git</a> version control system, providing a simplified but powerful web interface that allows users to participate in projects, contribute code, report and discuss software bugs, discover existing code and data, as well as publish new code.
Through version control, users have a detailed, chronological record on the files and directories stored in their repositories <span class="citation" data-cites="RVetqmsg">[<a href="#ref-RVetqmsg" role="doc-biblioref">6</a>]</span>.
This workflow raises a strong and clear advantage over receiving, processing and sending files back-and-forth, a process that can easily become challenging and time-consuming in projects extending in time and in the number of collaborators <span class="citation" data-cites="4ny1onB0">[<a href="#ref-4ny1onB0" role="doc-biblioref">7</a>]</span>.
Through the combination of version control management and the network- and collaboration-based features, GitHub can broadly facilitate openly available source code alongside concomitant collaborative development <span class="citation" data-cites="kEX5dgzK">[<a href="#ref-kEX5dgzK" role="doc-biblioref">8</a>]</span>.</p>
<p><a href="https://git-scm.com/">Git</a> is the version control system that enables all the collaborative tools available on GitHub.
In Git, versions of files and directories are uniquely identified as “commits”, allowing one to identify and track modifications line-by-line.
Although the understanding of basic concepts of Git (such as commit, push, pull, checkout; see (Box 1)[#box-1-definitions]) is necessary, the GitHub web-based platform and its integrated development environments (such as the GitHub Desktop) allow users to manage their repositories without using more technical command-line sessions.
We do not focus on Git in this paper, but we recommend users explore the many resources providing detailed information on Git, such as journal articles <span class="citation" data-cites="PlcxShQU kEX5dgzK">[<a href="#ref-kEX5dgzK" role="doc-biblioref">8</a>,<a href="#ref-PlcxShQU" role="doc-biblioref">9</a>]</span>, video tutorials, and books <span class="citation" data-cites="13jOlVcpp">[<a href="#ref-13jOlVcpp" role="doc-biblioref">10</a>/]</span>.</p>
<!-- ### General background on how GitHub is used by EEB researchers -->
<!-- *Contributors to this section: RCO, PHPB* -->
<p>The voluminous user-community and the numerous resources providing pedagogical instruction material on how to use GitHub streamlined its widespread use (CITE).
Nevertheless, although researchers in EEB have been encouraged to adopt GitHub as part of their research process <span class="citation" data-cites="3DKwn1sY 10ghgV3S8">[<a href="#ref-3DKwn1sY" role="doc-biblioref">11</a>,<a href="#ref-10ghgV3S8" role="doc-biblioref">12</a>]</span>, its use is still not widespread.
Because GitHub and its features have been centered on collaboration for software development in information systems <span class="citation" data-cites="139b0pSGc">[<a href="#ref-139b0pSGc" role="doc-biblioref">13</a>]</span>, first-time users from domains without formal training in information technology may face steep learning curves.
Moreover, domain-specific perspectives and resources providing tractable examples and practical guidance for researchers in ecology and evolution on GitHub are scarce (but see https://ourcodingclub.github.io/, https://www.openscapes.org/, ADD OTHER CITATIONS).
An increased availability of data and code management standards – of which GitHub is one increasingly important component – make research more reproducible and collaborative <span class="citation" data-cites="13QX8XU3J">[<a href="#ref-13QX8XU3J" role="doc-biblioref">14</a>]</span>.
More importantly, a widespread, common adoption of GitHub for collaborating on a variety of research tasks can ultimately enable EEB researchers spending less time on creating novel processes for collaboration and more time on their scientific research <span class="citation" data-cites="ydrk01SR">[<a href="#ref-ydrk01SR" role="doc-biblioref">15</a>]</span>.</p>
<!-- ### Our objective -->
<!-- *Contributors to this section: RCO, PHPB* -->
<p>This manuscript is the result from an academic hackathon held during the 2021 conference for the <a href="https://www.sortee.org/">Society for Open, Reliable, and Transparent Ecology and Evolutionary Biology (SORTEE)</a>.
We convened a group of ~30 EEB researchers with varying levels of familiarity with using GitHub as part of their research projects to showcase and discuss how existing features can contribute to the documentation and collaboration in ecological and evolutionary research.
During the hackathon, we identified the need for a formal discussion on how EEB researchers can benefit from GitHub and its features to make their research more collaborative and transparent.
Here, we outline thirteen practical ways that EEB researchers can use GitHub features for more collaborative, transparent, and reproducible science.
We also provide critical perspectives on features that could be improved and catered towards research development.</p>
<h3 id="box-1-definitions">Box 1: Definitions</h3>
<!-- Contributors to this section: ERS, Ali -->
<!--# I thought it might be helpful to have a box with short definitions of git/GitHub terminology used in the manuscript. If any of these are discussed more in depth in the main text, they may not need to be here. RCO: I think it's great. Super helpful to have these types of glossarys for github papers -->
<ul>
<li><p><strong>repository</strong>: A collection of files (e.g. a directory) tracked by git. Commonly shortened to “repo”</p></li>
<li><p><strong>commit</strong>: Commits are like snapshots in the development of a project.
Commits can include changes in multiple files and must include a brief commit message describing the changes made.
A typical workflow is to make some related changes in files, make a commit (e.g. “generate and include fig1 in results”), and after several commits <strong>push</strong> those commits to the remote GitHub <strong>repository</strong>.</p></li>
<li><p><strong>clone</strong>: Cloning a <strong>repository</strong> is a way of making a local copy (i.e. on your computer) of a GitHub <strong>repository</strong>.
If you have access to <strong>push</strong> to a <strong>repository</strong>, this can be a first step to contributing to a project.</p></li>
<li><p><strong>branch</strong>: Development branches can be created at any point in time and work on each branch can continue independently.
This is useful for testing out new ideas (both code and text) which may or may not eventually get integrated into the main branch of the project.
Branches can also be used to isolate contributions of multiple contributors.
Each person working on their own branch eliminates problems that arise when conflicting edits are pushed to the same remote branch.
Changes in a development branch can be <strong>merged</strong> into the main branch via <strong>pull requests</strong>.
Branches can only be made by those who are given access to the project <strong>repository</strong>.</p></li>
<li><p><strong>fork</strong>: A fork is a copy of a <strong>repository</strong> hosted on GitHub.
If a repository is public, then anyone can make a fork.
Even if they do not have access to push to the original repository, they can make a fork and edit it independently.
Forks are linked to the original GitHub repository and “upstream” changes (those in the original repository) can be <strong>merged</strong> to keep the fork up to date with the original project.
Changes made in the fork can be integrated into the original project via <strong>pull requests</strong>.</p></li>
<li><p><strong>push/pull</strong>: When <strong>commits</strong> are made in a project locally, they must be synced with the remote GitHub repository by “<strong>pushing</strong>” them.
Changes on a GitHub repository can then be “<strong>pulled</strong>” to keep your local version of the project up-to-date.</p></li>
<li><p><strong>pull request</strong>: A pull request is a request that the owner(s) of a GitHub repository integrate changes you’ve made on either a <strong>branch</strong> in the repository or in your own <strong>fork</strong>.
When you initiate a pull request, you must provide a description of what changes are made.
Some automated tests may be run and review may be required before integrating your changes into the main <strong>branch</strong>.</p></li>
<li><p><strong>merge</strong>: Combining <strong>commits</strong> from two different branches together into one <strong>branch</strong></p></li>
</ul>
<!-- ERS: while having a definition for all common git commands might be helpful to some, I think this box should probably only include jargon that gets used in the manuscript. So I think we should remove any definitions not used in the manuscript before submission
Ali: I agree with ERS and delete my additions except "merge".
RCO: Sounds good to me. This list is going to be super helpful, and we'll see if we need to remove any before submission. -->
<h3 id="box-2">Box 2</h3>
<!--TODO: add brief text describing table here -->
<figure>
<img src="images/box2.png" />
</figure>
<h2 id="github-in-ecoevo-examples-part-1">GitHub in EcoEvo examples (Part 1)</h2>
<h3 id="storing-and-archiving-version-controlled-data">Storing and archiving version-controlled data</h3>
<!--*contributors to this section: Dylan Gomes, Emma Hudgins* -->
<p>Many researchers often start their use of GitHub to backup their working data and code to a remote server (Just push and pull, see Box 1, from their own repo).
This saves the user time from backing up data and code on their own portable devices, such as hard drives.
This also offers some peace of mind, as this information is retrievable even if one’s laptop ends up at the bottom of a lake.
Thus, an additional benefit of this ‘cloud’ storage is that one’s GitHub repository can be accessed by any machine with internet access, allowing the user to be more mobile if they wish to both work from home and the office from different computers.
Each time a user pushes changes to their repository, GitHub tracks what these changes are and stores this history.
This feature allows for version control, such that users can re-visit previous versions of data and code.
Version control is particularly useful if a mistake has been made where a user has unknowingly overwritten or deleted information that would otherwise be irretrievable without GitHub having saved that information.</p>
<p>An even easier way to start using Github is for the archival of cleaned code and data, often accompanying preprinting, manuscript submission, or manuscript acceptance.
Many users prefer to host a separate, cleaned repository that they make public when they complete a paper, while keeping the original folders as either a private GitHub repository, or on another cloud storage service such as OneDrive, Dropbox, etc.
One benefit of using GitHub for this service is that it can integrate with a website called <a href="www.zenodo.org">Zenodo</a>, a free, long-term data archiving service funded by CERN.
After linking your GitHub account to Zenodo and turning on archiving, any time a release (Box 1) is made, a snapshot of the entire repository is archived in Zenodo with a versioned, citable DOI (see <code>### Making code citable</code> below for more information).
DOI’s for data and code are increasingly being required by journals for paper acceptance (e.g., Journal of Applied Ecology), and Zenodo provides a free alternative to other fee-based hosting services (such as Dryad).</p>
<h3 id="virtual-lab-notebook">Virtual lab notebook</h3>
<!--*contributors to this section:* -->
<p>commits as a way to record daily progress
issues as a way to keep track of short-term objectives/goals, and progress towards them</p>
<h3 id="classroom-teaching-educational-materials">Classroom teaching / educational materials</h3>
<!-- *contributors to this section: Cole Brookson* -->
<p>GitHub provides a large variety of uses for hosting teaching/educational materials.
In fact, through taking advantage of the suite of GitHub features, the entire process of running a course, workshop, or even just a lecture, can all be done openly on GitHub.<br />
As a matter of gross simplification, organizing a course (for example) could be broken down into: 1) developing the material (i.e., slides, examples, relevant readings, labs, etc.), 2) hosting the course on some online platform for students to access, 3) delivering the content, and 4) accepting student work submissions and then returning graded material.<br />
While of course there are other purpose-built platforms for this type of activity, few of them provide the usability at the price point GitHub does.</p>
<p>First, developing course material, from slides to labs and everything in between, can be done on GitHub, out in the open, where others can see, review and offer feedback on your process!<br />
Making presentations can be done through most major high-level programming languages such as <a href="https://rmarkdown.rstudio.com/lesson-11.html">R, with <code>RMarkdown</code></a>, <a href="https://python-pptx.readthedocs.io/en/latest/index.html">Python, with <code>python-ppt</code></a>, and <a href="https://juliapackages.com/p/remark">Julia, with <code>Remark.jl</code></a>.<br />
Since all these programs work via code bases, they can be version-controlled through git and GitHub.<br />
Once you’ve made all the content for your course, hosting a course website can be done through GitHub pages, and there are <a href="https://github.com/topics/course-website">lots of templates available</a>.<br />
This way, not only can the course content be available to your enrolled students, but also to a global pool of learners and teachers interested in the course material.<br />
Since the course material can be easily housed on a GitHub pages website, it is then simple enough to deliver the content via that website, and/or a GitHub organization with template repositories for assignments etc.<br />
Student submissions are perhaps the least seamless component, but for assignments submitted as code files (i.e., <code>.R</code> &amp; <code>.Rmd</code> as two of the most common) and/or <code>.pdf</code> files, GitHub has a new and far-from-perfect but still useful tool <a href="https://classroom.github.com/">GitHub classroom</a> where instructors can host private assignments, and even build custom autograding tests, that will autograde assignments!</p>
<p>The previous section is meant to highlight the myriad tools GitHub can provide to centralize the delivery of educational materials.
While most instructors will likely choose to pick from this selection and end up having a mix of tools to deliver their content to students, it is still valuable to utilize some of these, if only for the reason that it can encourage students to even <em>begin</em> learning about version control through interacting with git/GitHub, however minimally, through the course.<br />
There are (as always) no “points” awarded for using ALL GitHub materials ALL the time, but if a central tenant of a given course or educational unit is to introduce or give students experience to version control and the tools that working professionals in EEB use, then adopting a few of these tools can be a great way to do so.</p>
<!-- May want to incorporate the below citation into section on GitHub for teaching-->
<!--
Matthew D. Beckman, Mine Çetinkaya-Rundel, Nicholas J. Horton, Colin W. Rundel, Adam J. Sullivan & Maria Tackett (2021) Implementing Version Control With Git and GitHub as a Learning Objective in Statistics and Data Science Courses, Journal of Statistics and Data Science Education, 29:sup1, S132-S144, DOI: 10.1080/10691898.2020.1848485 -->
<h2 id="github-in-ecoevo-examples-part-2">GitHub in EcoEvo examples (Part 2)</h2>
<h3 id="project-management">Project management</h3>
<p><em>Contributors to this section: Kaitlyn Gaynor, Rob Crystal-Ornelas</em></p>
<p>GitHub can be a powerful tool for team-based project management, allowing collaborators to share feedback, brainstorm ideas, and troubleshoot problems (Figure 2).
The “Issues” feature of GitHub allows for discrete tasks and sub-tasks to be identified, assigned to team members, and categorized with custom labels.
The new GitHub “Discussion” feature serves as a message board for conversation.
Scripts, commit messages, and pull requests can be linked directly to issues and discussions, providing a clear record of project workflow.
The use of GitHub for all project-related conversation and planning, rather than e-mail or messaging tools, makes it easier to keep track of progress throughout the lifespan of a project.
This is because unlike emails and messages which can get lost as more new tasks arise, GitHub issues exist until they are intentionally closed by repository administrators.
Fortunately, it is not essential for all team members to have proficiency in git or programming, as users can interact with Issues and Discussions via web browser or e-mail (e-mail responses still get tracked as comments on the focal GitHub issue).
For larger projects with many team members and tens or hundreds of GitHub issues to sort through, project management software like ZenHub, can help prioritize issues and pull requests.
ZenHub’s web interface includes a GitHub Issue visualizer where users can organize issues into high priority or backlogged tasks as well as link issues togther when they are related to a shared project goal or milestone.
GitHub is currently beta testing a similar project management feature called <a href="https://docs.github.com/en/issues/trying-out-the-new-projects-experience/about-projects">GitHub Projects</a>. GitHub can bGitHub can also be integrated with other project management software like Slack or Zenhub.</p>
<h3 id="building-website">Building website</h3>
<p><em>Contributors to this section: Rob Crystal-Ornelas, Emma Hudgins</em><br />
It is now common for many scientists to have personal, project, or lab websites (hereafter, personal websites) to share and promote their work.<br />
There are many options for creating and hosting websites.<br />
Some sites are built through a point-and-click user interface that requires no coding experience, but these services tend to have monthly or annual fees (e.g., Wix, Squarespace, Wordpress).<br />
<a href="https://pages.github.com/">GitHub Pages</a> allows users with a GitHub account to easily create a website, hosted by GitHub, from one of their many website templates <span class="citation" data-cites="10ghgV3S8">[<a href="#ref-10ghgV3S8" role="doc-biblioref">12</a>]</span>.<br />
It is also possible to fork any public website hosted on GitHub in order to use it as a template.<br />
When creating a website with GitHub Pages, all content is stored in a GitHub repository, the content is written in markdown (e.g., https://github.com/SORTEE-Github-Hackathon/main-website), and a website is automatically rendered in HTML from the markdown documents (e.g., https://sortee-github-hackathon.github.io/main-website/).<br />
Aside from free hosting services, another benefit is that GitHub pages are autogenerated, meaning that when content is modified in the associated GitHub repository, the website instantly updates <span class="citation" data-cites="RVetqmsg">[<a href="#ref-RVetqmsg" role="doc-biblioref">6</a>]</span>.<br />
Though the templates are useful for quickly starting up a new website, users are able to fully customize their Pages websites (for technical details of customizing GitHub Pages site see <span class="citation" data-cites="MXxgZJ45">[<a href="#ref-MXxgZJ45" role="doc-biblioref">16</a>]</span>).<br />
We emphasize that despite the many benefits of using GitHub pages (free hosting, templates, customization), this avenue for creating a website will be more time intensive than the out of the box platforms mentioned above and requires consideration of tradeoffs offered by website creation services.<br />
For more advance GitHub users, <a href="https://jekyllrb.com">Jekyll</a> and <a href="https://gohugo.io">Hugo</a> are both “static website generators”, which also include template libraries for websites that can be hosted freely via GitHub pages.
Both of these tools require some additional learning because they are deployed via the computer’s terminal or command line, still they are a great resource for creating free, eye-catching websites.</p>
<h3 id="making-code-citable">Making code citable</h3>
<p><em>Contributors to this section: Rob Crystal-Ornelas, Emma Hudgins, Dylan Gomes</em><br />
GitHub makes it easy to store and share a variety of data files in the cloud.
If a repository is made “public” the URL to the repository can be shared freely with others.
However, for a variety of reasons (e.g., privately owned company, ability to make repositories private, accounts can be deleted at will) GitHub is not considered a long-term data or code repository like <a href="https://zenodo.org/">zenodo</a> and <a href="https://figshare.com/">figshare</a> <span class="citation" data-cites="10ghgV3S8 kEX5dgzK">[<a href="#ref-kEX5dgzK" role="doc-biblioref">8</a>,<a href="#ref-10ghgV3S8" role="doc-biblioref">12</a>]</span>.
Also, unlike the long-term repositories, GitHub does not issue Digital Object Identifiers (DOIs) for content uploaded to their servers.
DOIs are persistent and unique alpha-numeric IDs assigned to research products like papers, code, and data.
DOIs allows tracking and citing research products.
For this reason, scientists who share code and data through GitHub are strongly encouraged to also submit GitHub repository content to a long-term data archive <span class="citation" data-cites="1Du6fzB8g">[<a href="#ref-1Du6fzB8g" role="doc-biblioref">17</a>]</span>.
Fortunately, both long-term repositories mentioned above (Zenodo and Figshare) have integrations with GitHub which facilitates archiving a snapshot of all repository content with the click of a button.</p>
<p>Linking one’s GitHub repository with Zenodo, etc. to achieve a DOI helps work become findable, gives proper attribution, and that can ensure long-term stability (Hampton et al. 2015).
Thus, when researchers wish to include data and code with their publications, they ought to reference a DOI from a long-term storage site, rather than a URL from GitHub (which can change or be deleted).
Additionally, referencing a DOI for data and code is preferable to submitting these as supplementary materials to the journal, as supplementary materials are more difficult to find and reuse (i.e. often not centralized and searchable in a database) and not necessarily permanent (as most journals offer no guarantee of long-term storage).</p>
<p>Many researchers believe that their code is not useful because their analysis is context-specific and not designed for re-use like software.
However, there are many reasons to share data and code beyond re-use.
Even if code is rough, it shows the exact steps taken to conduct an analysis, and therefore provides the most detailed look into how to reproduce a given analysis <span class="citation" data-cites="uBJwnPbq">[<a href="#ref-uBJwnPbq" role="doc-biblioref">18</a>]</span>.
This is important in light of the reproducibility crisis [https://doi.org/10.1038/533452a] and will become increasingly important to the collective scientific enterprise as advances in computing power and accessibility unlock the ability to conduct ‘big data’ meta research with data that has already been collected by others.
Failing to include data and code with our publications leaves future scientists with many fewer resources from which to understand the world.</p>
<p>The standard GitHub licensing options are best suited for software.
If your code is intended only for your specific analysis, consider a Creative Commons License.
The <a href="https://choosealicense.com/non-software/">Choose a License</a> website can offer further guidance.
If you wish to allow anyone to re-use your code, consider a CC0 1.0 public domain dedication.
If you wish to receive attribution for any reuse of your code, consider a CC BY 4.0 license, which requires attribution upon reuse.
If you have build an app, tool, package, or other product that you would like others to use and would like attribution for any reuse of your code, consider the GNU General Public License v3.
This license also prohibits the re-user from making their re-used version private.
If you do not wish to receive attribution and are open to private use, consider the MIT license.</p>
<h3 id="collaborative-code-editing">Collaborative (code) editing</h3>
<p><em>Contributors to this section: Kaitlyn Gaynor, Rob Crystal-Ornelas</em></p>
<p>From its inception, one of the primary uses of GitHub has been for collaborative coding.
We acknowledge that there are important differences between the average software developer and ecology/evolution researcher using GitHub, and that not all GitHub collaboration features are optimal for research purposes.
However, core features of git like forking and branching can allow for simultaneous coding on different versions of the same research project, and alternative versions can be easily discussed and resolved with GitHub.
While a complete review of these features is beyond the scope of our paper, there are many free resources for learning how to use these collaborative features of GitHub <span class="citation" data-cites="WrDwZxcM">[<a href="#ref-WrDwZxcM" role="doc-biblioref">19</a>]</span>. (e.g. <a href="https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request" class="uri">https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request</a>)
It is often best to develop comfort with features like pull requests and merges on “practice” repositories with colleagues before integrating these tools fully into a collaborative workflow.</p>
<p>GitHub can also facilitate interactions between research advisors and advisees, providing a platform for students or other trainees to share in-progress code, and flag specific challenges or questions for their supervisors or mentors.
Periodic code review can also help advisors to identify errors early in the process, and inform further training and mentorship to fill gaps in skills.</p>
<h3 id="writing-manuscript">Writing manuscript</h3>
<p><em>contributors to this section:</em><br />
Caveat that GitHub has been called out for not being so user-friendly for manuscript development (Ram 2013).
But getting better?
Tools that link with GitHub have been developed with synchronous writing in mind.
HackMD provides a collaborative writing platform based on Markdown that integrates with GitHub.<br />
We used this platform early on in the process of writing this manuscript to generate an outline.</p>
<p>Later, we wrote the manuscript by Manubot, which is one of the important collaborative manuscript platform uses Markdown for writing and GitHub for storing and tracking changes over time <span class="citation" data-cites="nxqXElqz">[<a href="#ref-nxqXElqz" role="doc-biblioref">20</a>]</span></p>
<p>Recently, more manuscripts were written on GitHub via Manubot(see examples <a href="https://manubot.org/catalog/" class="uri">https://manubot.org/catalog/</a>)</p>
<h3 id="peer-review">Peer-Review</h3>
<!--*contributors to this section:* Eric R. Scott-->
<p>Peer review of research software by rOpenSci (<a href="https://ropensci.org/software-review/" class="uri">https://ropensci.org/software-review/</a>) and of research software and associated manuscripts by the Journal of Open Source Software (<a href="https://joss.readthedocs.io/en/latest/submitting.html" class="uri">https://joss.readthedocs.io/en/latest/submitting.html</a>) requires that submitted work is hosted on GitHub and their review processes make use of GitHub issues (Box 1).
<!--# not sure if links are really necessary -->GitHub can also be used as a hub for reviewers and authors during the peer review process of an ordinary research manuscript.
If the code associated with a manuscript is made available at the time of submission (e.g. via a link to a GitHub repository in a Data Availability Statement), peer-reviewers may be able to offer more helpful suggestions on written methods and may even make comments on the code itself, potentially catching bugs or errors before publication <!--# would be nice to have an example to link to here.  I suspect it is rare that reviewers look at code, but it's happened to me (ERS) -->.
GitHub issues (Box 1) can also be used to organize and discuss reviewer suggestions and to assign them to co-authors (See example <a href="https://github.com/BrunaLab/HeliconiaDemography/issues?q=is%3Aissue+label%3A%22reviewer+comment%22+">here</a>).
When reviewer comments are posted as separate issues, authors can comment on the issues to discuss possible changes and assign themselves to indicate which comments they intend to handle.
Co-authors can then integrate their edits and responses to reviewers using pull requests (Box 1).</p>
<h2 id="github-in-ecoevo-examples-part-3">GitHub in EcoEvo examples (Part 3)</h2>
<h3 id="open-science-discussion">Open science discussion</h3>
<p><em>Contributors to this section: Freddy Hillemann, Allison Binley</em>
Research papers are condensed outputs that hide the underlying intellectual and computational workflows, including the treatment of the raw data and analytical steps.
Granting readers access to code and other documentation of the analysis allows them to retrace and comprehend analytical decisions.
Github provides a platform to access all aspects of the project, using citable DOIs, rather than just the final manuscript.
While often thought of as storage for data and code, github repositories can also be used to publish a time-stamped preregistration of research plans and hypotheses.</p>
<p>Github is a tool for managing and sharing components of any research project, that can help accelerate progress towards Open Science goals <span class="citation" data-cites="1A97a4UwU">[<a href="#ref-1A97a4UwU" role="doc-biblioref">21</a>]</span>, from developing the analysis through to publication and ensuring reproducibility.
Conventional research practices typically rely on one or two people running and checking the data analyses, while most coauthors (and readers of the subsequent publication) see only the final results and a verbal description of the analytical steps.
In the developmental stages, collaborators can directly see the code for the analysis, manipulate and explore the data themselves, and check for errors. Cynically, there is also more insurance against nefarious colleagues that may be tempted to distort <a href="https://ecologyforthemasses.com/2020/02/04/pruittdata-and-the-ethics-of-data-in-science/">results</a>.
Collaborators are better positioned to discover questionable findings if they have full and transparent access to the project.</p>
<p>This transparency can similarly be extended beyond coauthors to the entire scientific community.
Publishing the data and reproducible workflows along with the manuscript allows any reader to review the analysis and reproduce the experiment <span class="citation" data-cites="doi.org/10.1371/journal.pbio.3000763">[<a href="#ref-doi.org/10.1371/journal.pbio.3000763" role="doc-biblioref"><strong>doi.org/10.1371/journal.pbio.3000763?</strong></a>]</span>.
Supplying code for (novel) methods that are proposed or used also reduces barriers to knowledge and can greatly improve the ability of others to build on existing work, resulting in greater proliferation and accessibility for a broader audience.
Github even provides a useful <a href="https://docs.github.com/en/discussions">Discussions Forum</a> that aids the direct communication with repository owners, as well as the <a href="https://github.community/">Github Community</a> forum for more general questions and sharing of expertise.</p>
<p>If the methods and analyses used are fully available during the peer-review process, they can be as critically evaluated as the rest of the manuscript. <em>link to peer review section?</em>.
However, fear of being scooped feeds into cultural reluctance to do science openly due to worries of intellectual property.
A solution would be to allow sharing a private link with a citable DOI for peer review, and update the DOI post-acceptance or post-publication.
Updating DOIs is currently yet a missing feature of github, but implemented in other open-access repositories such as Zenodo, which was developed under the European OpenAIRE program and is being hosted by CERN.
Alternatively, github users can to keep their data and code repository private until publication, and then make it public.</p>
<h3 id="project-continuity">Project continuity</h3>
<p><em>Contributors to this section: BPME, VF</em></p>
<p>The development of research software continues to be on the rise, and with that comes the need to consider the continuity of the research software.
This is particularly relevant for software developed for relatively short-term research projects, such as projects developed by graduate students or postdoctoral fellows <span class="citation" data-cites="D4C4k4ak">[<a href="#ref-D4C4k4ak" role="doc-biblioref">22</a>]</span>.
Often with these projects, once the research contract expires, the research software upkeep tends to fall off as the researchers move on to new projects.
Additionally, if the research software is kept on only the researcher’s hard drive, it becomes increasingly difficult to access the software and code for future uses.</p>
<p>When the project owner is finished with the project, or their contract expires, there generally should be a handover period of this software in order for the next cohort of researchers to reuse what was already developed [<span class="citation" data-cites="D4C4k4ak">[<a href="#ref-D4C4k4ak" role="doc-biblioref">22</a>]</span>]<span class="citation" data-cites="4ny1onB0">[<a href="#ref-4ny1onB0" role="doc-biblioref">7</a>]</span>.
GitHub facilitates project continuity among research software and research code by providing tools that make this handover period easier.
As we have already mentioned, using Git for code in Ecology and Evolution can allow for a “paper trail” of sorts to be created for the research software, thus allowing for future owners of the code access to the entire history of the project <span class="citation" data-cites="3DKwn1sY">[<a href="#ref-3DKwn1sY" role="doc-biblioref">11</a>]</span>.
Additionally, GitHub allows for repositories and organizations to have designated Code Owners <span class="citation" data-cites="s91uGRZ2">[<a href="#ref-s91uGRZ2" role="doc-biblioref">23</a>]</span>; these code owners can change through time allowing for the transition of research software from one cohort of researchers to the next <span class="citation" data-cites="iIEKCTLU">[<a href="#ref-iIEKCTLU" role="doc-biblioref">24</a>]</span>.</p>
<p>Within EEB projects, tasks are often divided among contributors taking various roles (see <a href="https://casrai.org/credit/">CRediT taxonomy</a>).
The creation of project repositories is commonly the purview of those involved in the software, formal analysis, and/or visualization components of the project through their roles as code writers.
However, the structural components of a typical GitHub repository and the derived EEB-specific templates can provide functional ways for other non-code writers to be engaged in aspects of repository design in a way that improves institutional memory and facilitates project continuity.
Non-code writers can offer many contributions to repository design and development, and their active involvement can both aid authors ability to act as guarantors of the project, and the clarity and reproducibility of the project for future users.
In Figure 2, we highlight several elements of good repository structure, and the various ways that contributors may interact with them.</p>
<h3 id="asynchronous-working">Asynchronous working</h3>
<p><em>Contributors to this section: Ali, Allison Binley</em></p>
<p>Recently, asynchronous communication tools were boosted the team works.
Github served as an excellent environment for asynchronous communication and collaboration for especially remote team projects.
Researchers can easily collaborate without being in same place and time.</p>
<p>One of the most useful aspects of Github is its propensity to facilitate remote and asynchronous collaboration.
Researchers can seemlessly access and contribute to data and code regardless of disparities in schedules or location.
This is particularly important given the increase in remote work in recent years, but the benefits can also extend far beyond the “work from home” model.
Improving remote collaboration can encourage the exchange of ideas among researchers at different institutions and in different countries, which can serve to improve the quality of the research itself.
Researchers can work directly with experts from all over the world, who have access to the same data and code as they do.</p>
<p>However, Github has something to offer even for team members on a project that work in the same office.
Researchers can easily stay abreast of progress made by other collaborators without the need for meetings or emails.
Collaborative project work can also be clearly split between team members, giving them the flexibility to contribute when it best fits their schedule.
The <a href="https://sortee-github-hackathon.github.io/manuscript/v/latest/index.html#storing-and-archiving-version-controlled-data">version control</a> features also allow users to make progress and changes without worrying about irreparably writing over someone else’s work.</p>
<p><em>could link this back to the “Project Management” section or even “Collaborative (code) editing”.It actually seems to me like this entire section could potentially be combined with “Collaborative (code) editing”</em></p>
<h3 id="github-organizations">GitHub organizations</h3>
<p><em>Contributors to this section: Katherine Hébert, Cole Brookson</em></p>
<p>Whether experiments are done in a wetlab, data are gathered in a field site, or analyses are run in a shared office, even conceptually distinct projects are often carried out in a common physical space.
GitHub Organisations offer a shared virtual space that allows a team to work in different repositories, while remaining tied together under a larger figurehead, such as a laboratory, a department, an organisation, or a large project involving several teams.
Organisations are well-suited to ensure larger projects with many steps or moving parts are constrained to one virtual space, where outputs and sub-projects can be easily accessed and located without relying on any one individual.
Because the repositories are grouped in one virtual space, members can reference and contribute to each other’s work without necessarily being part of the same repository, broadening the accessibility and longevity of code and writing contributions.</p>
<p>Contributors can be assembled into teams within an organisation, which allows administrators to assign roles and tasks to groups of people.
Whereas access to repositories is usually assigned to individual contributors, Organizations facilitate the management of access permissions by allowing each team to be granted access to certain repositories, and not to others.
This ensures that more sensitive repositories remain as restricted as needed, while repositories with greater general interest can be easily accessible to many members at once.</p>
<p>As an example, GitHub Organizations are particularly well-suited to house documents and projects within a laboratory, such as research compendia, codes of conduct, protocols, training documents, and other such documents that evolve collaboratively over time and are relevant to many colleagues.
In this way, students or teams can have full ownership of repositories within an organization, while ensuring that these materials stay accessible to the laboratory after people have moved on (or upgraded their computers).
This application extends to research centres, which may include several distinct projects that remain linked under a given institution, such as the <a href="https://github.com/idiv-biodiversity">German Centre for Integrative Biodiversity Research (iDiv)</a>.
Of course, the utility of this tool goes beyond laboratories - they are useful to structure the organisation, presentation, and outcomes of working groups such as the hackathon which inspired this paper (<a href="https://github.com/SORTEE-Github-Hackathon">SORTEE-Github-Hackathon</a>) by keeping track of all materials as ideas develop and take shape in one virtual space.
Organisations are also convenient for hosting a set of related learning materials such as a set of lectures or workshops, such as the Québec Centre for Biodiversity Science R Workshop Series (<a href="https://github.com/QCBSRworkshops">QCBSRworkshops</a>) or the University of Edinburgh’s Coding Club (<a href="https://github.com/ourcodingclub">Coding Club</a>), which may be updated by an ever-evolving group of contributors over time.</p>
<p>Utilizing GitHub organizations as a research group or even for a handful of individuals working on a group of projects can be incredibly useful for all involved.
GitHub organizations are relatively easy to set up, and especially easy to manage as membership to the organization changes through time.
Not only is it a useful way to store repositories of lab-related research products, but it’s also incredibly helpful for storing “living documents” that may be edited frequently, and may be linked to a lab website (that could also be generated via a repository that lives within the organization!).
The use of the “Teams” feature can allow certain groups to have varying levels of access to repos in the organization with a select group having push access to some repos but not others.
This can manifest in a group working on some common dataset(s) (e.g. some genetic data) to have push access to the handful of repositories used for processing sequence data, while another group of students/researchers may have push access to an entirely different set of repos.
The organization structure also allows for easy tracking of issues, projects, and discussions related to the research group, and provides PIs/group leads an easy birds-eye view of the progress going on across multiple projects.</p>
<p>As well, organizations provide a convenient location for students to archive the code for their projects, for use/reference by future students in the research group, thus providing a type of knowledge communication that may not exist otherwise.
Indeed, providing new students with access to the organization and ideally a template repository for lab projects can soften the burden on those new to the software, in that it provides them with examples to work off of, and an online location to ask for help from their labmates and/or advisors through tools like projects, discussions, and issues.</p>
<h3 id="additional-uses-for-github-in-ecoevo-research">Additional uses for GitHub in EcoEvo research</h3>
<p><em>Contributors to this section: RCO, Ali</em></p>
<p>There are many more ways that EEB researchers can use GitHub for accelerating research collaborations, and we briefly highlight several here.<br />
First, there are increasing calls for ecological data to be more Findable, Accessible, Interoperable, and Reusable (FAIR) <span class="citation" data-cites="1GuRfvMeX 6DjakjNS">[<a href="#ref-1GuRfvMeX" role="doc-biblioref">25</a>,<a href="#ref-6DjakjNS" role="doc-biblioref">26</a>]</span>.<br />
A key component of data reusability is standardizing the ways (e.g., variable names, file formats) that research data are archived in long-term repositories.<br />
Recently, community-led data standardization efforts are taking place on GitHub <span class="citation" data-cites="1Du6fzB8g">[<a href="#ref-1Du6fzB8g" role="doc-biblioref">17</a>]</span>, where documents and templates can be version controlled and commented on by the user community <a href="https://github.com/ess-dive-community">e.g., ESS-DIVE’s GitHub Community Space</a>.</p>
<p>Ecologists who write code often use the R programming language, and the <a href="https://ropensci.org/">rOpenSci</a> community has a well-established software peer review process that involves both rOpenSci’s staff software engineers and the broader R user community.<br />
Their <a href="https://github.com/ropensci/software-review/">software review GitHub repository</a> provides instructions for submitting an R package for review as well as guidelines for code reviewers.<br />
rOpenSci’s efforts have resulted in many well-used R packages for ecology research including rfishbase <span class="citation" data-cites="Xsdcv6q">[<a href="#ref-Xsdcv6q" role="doc-biblioref">27</a>]</span> and taxize <span class="citation" data-cites="FVBWKkZu">[<a href="#ref-FVBWKkZu" role="doc-biblioref">28</a>]</span>.</p>
<p>Lastly, GitHub gists let users create and share snippets of code, notes, and files quickly.<br />
Rather than create an entire GitHub repository for saving a small code chunk you want to use in a presentation or share with a colleague, GitHub gists provide a lightweight way to write, save, and share code.<br />
Gists are associated with your Github account and can be public or private.<br />
Though gists lack all the features embedded in a GitHub repository, gists can still be forked, starred, downloaded, and easily added into a website or blog post.</p>
<h2 id="discussion">Discussion</h2>
<h3 id="general-pargrah-on-what-github-can-enable-in-ecoevo">General pargrah on what GitHub can enable in EcoEvo</h3>
<p><em>Contributors to this section: Rob, Brandon</em></p>
<p>There have been many calls for researchers outside of the software development community to join the 73 million GitHub users for their collaborative research.<br />
This call comes in light of the continual shift toward a more open-science framework across several fields, and as computational and data requirements increase in several fields.
Until now, resources and practical guidance specifically focused on using GitHub within the EEB community have been dispersed in blog posts and video tutorials.
We felt these resources have been extremely useful for us to learn to use Github in our own work, and we felt that a collation of the main ideas into one medium, while adding on our personal perspectives, would be of use in the EEB community.</p>
<p>In this paper, we describe 13 tractable ways that EEB researchers can leverage GitHub to enable more transparent and collaborative research (Figure 2).<br />
Many of the examples are specifically meant for first-time GitHub users and can likely be adopted with just several hours of practice (e.g., storing data, creating virtual notebooks, making code citable).
For example, storing code and data and making it citable generally just involve creating a repository on Github, pushing code to the repository, and then going through the necessary steps for creating a DOI for the repository.
These actions are generally covered in any introduction to Github tutorial and take little overhead to make it work.
On the other hand, some other examples we describe may require a greater time commitment, but have the potential to make EEB research even more open, accessible, and collaborative than ever before.
For example, managing full research projects or research labs on Github will require careful thought as to how to delegate tasks such as reviewing pull requests or creating issues, as well as thought as to how modular to make the research project or research lab (i.e., which repository will be used for what, and how many repositories are needed).
Additionally, collaboratively writing a manuscript using Github, such as what we have done with this, will involve a learning curve for folks less familiar with the intricacies of Github, and also require overheard to set up the repository using Github actions.
Despite the many potential applications of GitHub to EEB research, we acknowledge that there will still be many times when researchers might look to other platforms for research collaboration.</p>
<h3 id="when-to-look-to-other-platforms-for-collaboration">When to look to other platforms for collaboration</h3>
<p><em>Contributors to this section: Rob</em><br />
Though we see GitHub as a useful tool for collaboration in EEB, we describe 2 use cases where, to our knowledge, GitHub’s features still fall short of the type of highly collaborative work emblamatic of EEB research.<br />
First, because of the underlying version control system which tracks “pushed” changes through “commits”, real-time collaborative editing (e.g., as on a shared Google Doc or a Word document stored on Dropbox) is not possible on GitHub.<br />
There are now websites outside of the GitHub ecosystem that are built on top of the GitHub architecture that allow real-time collaboartive editing (e.g., <a href="https://hackmd.io/">hackMD</a>.<br />
We used HackMD at two key points in writing our manuscript when real-time co-writing was essential: when taking meeting notes and writing the outline of our paper.<br />
Despite the extra layer of coordination required to use GitHub to write a paper, many journals are changing their manuscript formatting requirements to allow for manuscript format output from GitHub <span class="citation" data-cites="AE0QcVgJ">[<a href="#ref-AE0QcVgJ" role="doc-biblioref">29</a>]</span>.<br />
Second, we looked to other software when working on figures and tables.
Creating tables and figures on GitHub using markdown or other scripting languages is possible, we found that it was not practical at the early brainstorming stages.
We needed to rapidly iterate on figure and table design, share feedback through comments, and merge/reorder ideas when necessary.
For these reasons, we used Google Slides for working on figures and Google Sheets for working on tables.
As our figures and tables moved toward more finalized forms, some co-authors chose to create the tables and figures using R and Markdown which could then be tracked using the same version control system as the rest of manuscript.</p>
<h3 id="why-arent-more-eeb-researchers-using-github">Why aren’t more EEB researchers using GitHub</h3>
<p><em>Contributors to this section: Saeed, Vivienne</em></p>
<p>Though GitHub has been available as a platform for more than a decade at time of writing, its uptake among EEB researchers, especially as a tool for collaboration, has been slow.
Some research groups choose to take advantage of alternative platforms with similar capabilities, such as GitLab or Bitbucket, but many have yet to integrate these types of project management software into their lab ecosystems at all.<br />
Some attribute this hesitation to the steep learning curve combined with limited instruction available through traditional university courses.
When use of GitHub is taught within an EEB context, it is usually accompanying coursework in topics such as statistical programming, and some students may find it overwhelming to juggle learning git alongside scripting languages, statistical theory, and file system navigation, especially when many may also be new to using command-line interfaces in general.
Instructors likewise may confuse the expected digital literacy of younger students with computational fluency, even when modern technology increasingly abstracts many relevant concepts through search optimization and preponderant IDEs.
The default public nature of GitHub usage can add additional pressure to students learning to use the platform.<br />
While many quantitative ecologists and evolutionary biologists take advantage of GitHub for individual use, collaborative use may lag due to how researchers traditionally divide labor within projects.
Despite broad utility, GitHub remains a tool predominantly used by computer scientists and software developers, and EEB researchers may take the view that GitHub is a platform that only needs to be used by individuals writing code, and may silo those aspects of projects to a single individual.<br />
Those assumptions may have obscured the utility of GitHub for tasks other than traditional data analysis and software development, or how GitHub can facilitate the integration of code with non-coding aspects of projects through the practice of repository design.
Additional hesitancy may come from general reluctance to share data and code publicly, or technical and logistical issues related to storage of large data files and lack of integration with other research platforms.</p>
<h3 id="our-own-limitations-in-writing-this-paper">Our own limitations in writing this paper</h3>
<p><em>Contributors to this section: Ali</em></p>
<p>Our own limitations since we are mostly writing from the EcoEvo perspective/ additional github limitation
Reliance on R since we are generally in EcoEvo
Discussion of free vs. paid plans. When projects get highly collaborative may have to add / pay for accounts. At this point, little difference between paid and free.</p>
<p>File size limitations make it hard to sync entire folders with the cloud, especially if your repo is private.
One trick to increase the size limit associated with a repository is to attach additional large files to the release once the repository is finalized and ready for re-use.</p>
<p>While we have written this manuscript collaboratively via GitHub and GitHub Actions, there was a substantial learning curve, and we are all early career researchers who are highly motivated to learn these skills.
This approach is likely less practical when collaborating with people at later career stages and a greater range of computational skills.
Real-time collaboration via GitHub is also not possible without relying on additional tools such as hackmd, so there is a longer delay to receive feedback from other collaborators, and a greater likelihood of conflicts.</p>
<p>Lack of GitHub help documents for non-English researchers in ecology and evolution leads them to miss the opportunity to fully understand the importance version control as well as the other benefits of GitHub.</p>
<p>Moreover. It seems there are sometimes unwritten roles that need to be explored in a way to learn and use GitHub and that in such Github collaborations, practice and redo practice play an important role. For instance, in one case during cloning the project to work on it, there will be unknown conflicts and limitations for downloading and cloning process on local computers because of low internet speed. But it seems it would be much easier if we choose “open with GitHub desktop” instead of using ” Download Zip</p>
<h3 id="using-github-is-a-good-start-but-lots-of-practices-to-make-repo-more-user-friendly">Using GitHub is a good start, but lots of practices to make repo more user friendly</h3>
<p><em>Contributors to this section: Ali, Emma</em></p>
<table>
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="header">
<th>10 Tips for getting started in GitHub</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>1. <strong>Check for an existing solution to your problem.</strong> The Github Help <a href="https://docs.github.com/en">webpage</a> contains extensive and detailed documents with helpful screenshots. It is a good starting point for handling an issue, and has troubleshooting tips for specific problems. Alternatively, consider Tweeting your issue. There is a large community of GitHub users around the world who have likely faced analogous problems and may be able to provide quick solutions. Third, try to follow blogs e.g.(https://github.blog/), Twitter accounts or YouTube channels that regularly post practical solutions about common Github issues.</td>
</tr>
<tr class="even">
<td>2- <strong>Consider taking free courses</strong> such as those from <a href="https://swcarpentry.github.io/git-novice/">Software Carpentries</a> and sharing these courses with your lab members or colleagues.</td>
</tr>
<tr class="odd">
<td>3. <strong>Take the advantage of Github as an asynchronous working tool for team-based projects</strong> See the repository for <a href="https://github.com/SORTEE-Github-Hackathon/manuscript/">this paper</a> as an example of a collaborative manuscript that includes discussions, issues, and a website via GitHub.</td>
</tr>
<tr class="even">
<td>4. <strong>The GitHub <a href="https://lab.github.com/">Learning Lab</a></strong> allows you to learn GitHub basics through short projects and tasks, and allows you to get feedback from their Learning Lab bot.</td>
</tr>
<tr class="odd">
<td>5. <strong>Check out the following <a href="http://markdownguide.org/basic-syntax/">markdown cheatsheet</a></strong> so that you can write clear metadata README files for your repositories.</td>
</tr>
<tr class="even">
<td>6. <strong>The Jenny Bryan universe of GitHub material</strong> provides a thorough and accessible introduction for a multitude of research-related uses for GitHub, and includes a <a href="http://happygitwithr.com">book</a>, <a href="http://stat545.com/">statistics course</a> and <span class="citation" data-cites="RVetqmsg">[<a href="#ref-RVetqmsg" role="doc-biblioref">6</a>]</span>.</td>
</tr>
<tr class="odd">
<td>7. <strong>Don’t be afraid of trial and error.</strong> One of the best ways to learn Github is the trial and error method. Learning from your own mistakes can be the better way to master your GitHub abilities. In any case, Github has the advantage of making it easy to go back to any steps that you desire via version controlling if you make mistakes.</td>
</tr>
<tr class="even">
<td>8. <strong>If you are an educator, include lectures on reproducibility and tools for creating reproducible workflows in your curricula</strong>. Some gradute programs now include coursework on course Rmarkdown and GitHub. Getting students started with these tools earlier will prevent the resistance that comes from working with a less reproducible workflow for a longer period of time.</td>
</tr>
<tr class="odd">
<td>9. <strong>Try to begin commiting with GUI (Graphical user interface) tools</strong> like <a href="https://desktop.github.com/">GitHub Desktop</a>, <a href="https://git-scm.com/docs/git-gui">git-gui</a>, <a href="https://www.rstudio.com/">RStudio</a>, <a href="https://code.visualstudio.com/">Visual Studio Code</a>, <a href="https://atom.io/">Atom</a>, <a href="https://www.gitkraken.com/">GitKraken</a> tools instead CLI (Command line interface) tools such as Terminal or Console for more advanced features.</td>
</tr>
<tr class="even">
<td>10. <strong>Get help deciphering GitHub Notifications.</strong> Try using tools like <a href="https://octobox.io/">Octobox</a> to disentangle and manage multiple notifications from distinct GitHub projects.</td>
</tr>
</tbody>
</table>
<h2 id="conclusion">Conclusion</h2>
<p>We provide 13 practical ways that Ecologists and Evolutionary Biologists can incorporate GitHub into their workflows.<br />
GitHub is still an emerging platform for many working in EEB, and so we include definitions (Box 1) and key user groups (Box 2) that can help researchers prioritize which GitHub skills to learn first.<br />
Some of the ways we outline for using GitHub are highly collaborative (e.g., open science discussion and collaborative code editing) while others are focused on individual actions (e.g., storing code/data, building a website).<br />
Regardless of the degree of immediate collaboration, adoption of GitHub in Ecology and Evolution has the potential to make the field more open and transparent than ever before.<br />
Our paper provides the most comprehensive review of applications of GitHub within EEB to date, and we encourage EEB researchers at any career stage studying any topic to try GitHub as a platform for sharing and collaboration.</p>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>This manuscript arose from a hackathon at the Society for Open, Reliable, and Transparent Ecology and Evolution (SORTEE) virtual meeting in 2020.</p>
<p>RCO was funded by the U.S. Department of Energy, Office of Science, Office of Biological and Environmental Research, Earth and Environmental Sciences Division, Data Management program under contract number DE-AC02-05CH11231.</p>
<h2 id="code-and-data-availability">Code and data availability</h2>
<p>The source code and data for this manuscript are available at https://github.com/SORTEE-Github-Hackathon/manuscript.</p>
<h2 class="page_break_before" id="references">References</h2>
<!-- Explicitly insert bibliography here -->
<div id="refs" class="references csl-bib-body" role="doc-bibliography">
<div id="ref-fJWFe93e" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">1. </div><div class="csl-right-inline"><strong>How do scientists develop and use scientific software?</strong> <div class="csl-block">Jo Erskine Hannay, Carolyn MacLeod, Janice Singer, Hans Petter Langtangen, Dietmar Pfahl, Greg Wilson</div> <em>2009 ICSE Workshop on Software Engineering for Computational Science and Engineering</em> (2009-05) <a href="https://doi.org/bw966x">https://doi.org/bw966x</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/secse.2009.5069155">10.1109/secse.2009.5069155</a></div></div>
</div>
<div id="ref-1Kqna6l2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">2. </div><div class="csl-right-inline"><strong>Challenge to scientists: does your ten-year-old code still run?</strong> <div class="csl-block">Jeffrey M Perkel</div> <em>Nature</em> (2020-08-24) <a href="https://doi.org/gg89cr">https://doi.org/gg89cr</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/d41586-020-02462-7">10.1038/d41586-020-02462-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32839567">32839567</a></div></div>
</div>
<div id="ref-10SpoByIw" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">3. </div><div class="csl-right-inline"><strong>A survey of the practice of computational science</strong> <div class="csl-block">Prakash Prabhu, Yun Zhang, Soumyadeep Ghosh, David I August, Jialu Huang, Stephen Beard, Hanjun Kim, Taewook Oh, Thomas B Jablin, Nick P Johnson, … David Walker</div> <em>State of the Practice Reports on - SC '11</em> (2011) <a href="https://doi.org/bdpsrp">https://doi.org/bdpsrp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1145/2063348.2063374">10.1145/2063348.2063374</a></div></div>
</div>
<div id="ref-O6UbstGG" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">4. </div><div class="csl-right-inline"><strong>Best Practices for Scientific Computing</strong> <div class="csl-block">Greg Wilson, DA Aruliah, CTitus Brown, Neil P Chue Hong, Matt Davis, Richard T Guy, Steven HD Haddock, Kathryn D Huff, Ian M Mitchell, Mark D Plumbley, … Paul Wilson</div> <em>PLoS Biology</em> (2014-01-07) <a href="https://doi.org/qtt">https://doi.org/qtt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pbio.1001745">10.1371/journal.pbio.1001745</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24415924">24415924</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3886731">PMC3886731</a></div></div>
</div>
<div id="ref-nwCtHDCn" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">5. </div><div class="csl-right-inline"><strong>Build software better, together</strong> <div class="csl-block">GitHub</div> <a href="https://github.com">https://github.com</a></div>
</div>
<div id="ref-RVetqmsg" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">6. </div><div class="csl-right-inline"><strong>Excuse Me, Do You Have a Moment to Talk About Version Control?</strong> <div class="csl-block">Jennifer Bryan</div> <em>The American Statistician</em> (2018-01-02) <a href="https://doi.org/gdhzdp">https://doi.org/gdhzdp</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1080/00031305.2017.1399928">10.1080/00031305.2017.1399928</a></div></div>
</div>
<div id="ref-4ny1onB0" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">7. </div><div class="csl-right-inline"><strong>Git can facilitate greater reproducibility and increased transparency in science</strong> <div class="csl-block">Karthik Ram</div> <em>Source Code for Biology and Medicine</em> (2013-02-28) <a href="https://doi.org/krv">https://doi.org/krv</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1186/1751-0473-8-7">10.1186/1751-0473-8-7</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23448176">23448176</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3639880">PMC3639880</a></div></div>
</div>
<div id="ref-kEX5dgzK" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">8. </div><div class="csl-right-inline"><strong>Ten Simple Rules for Taking Advantage of Git and GitHub</strong> <div class="csl-block">Yasset Perez-Riverol, Laurent Gatto, Rui Wang, Timo Sachsenberg, Julian Uszkoreit, Felipe da Veiga Leprevost, Christian Fufezan, Tobias Ternent, Stephen J Eglen, Daniel S Katz, … Juan Antonio Vizcaíno</div> <em>PLOS Computational Biology</em> (2016-07-14) <a href="https://doi.org/gbrb39">https://doi.org/gbrb39</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pcbi.1004947">10.1371/journal.pcbi.1004947</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27415786">27415786</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4945047">PMC4945047</a></div></div>
</div>
<div id="ref-PlcxShQU" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">9. </div><div class="csl-right-inline"><strong>A Quick Introduction to Version Control with Git and GitHub</strong> <div class="csl-block">John D Blischak, Emily R Davenport, Greg Wilson</div> <em>PLOS Computational Biology</em> (2016-01-19) <a href="https://doi.org/gbqsnf">https://doi.org/gbqsnf</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pcbi.1004668">10.1371/journal.pcbi.1004668</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26785377">26785377</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4718703">PMC4718703</a></div></div>
</div>
<div id="ref-13jOlVcpp" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">10. </div><div class="csl-right-inline"><strong>Let’s Git started | Happy Git and GitHub for the useR</strong> <div class="csl-block">Jenny Bryan Hester the STAT 545 TAs, Jim</div> <a href="https://happygitwithr.com/">https://happygitwithr.com/</a></div>
</div>
<div id="ref-3DKwn1sY" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">11. </div><div class="csl-right-inline"><strong>Our path to better science in less time using open data science tools</strong> <div class="csl-block">Julia SStewart Lowndes, Benjamin D Best, Courtney Scarborough, Jamie C Afflerbach, Melanie R Frazier, Casey C O’Hara, Ning Jiang, Benjamin S Halpern</div> <em>Nature Ecology &amp;amp; Evolution</em> (2017-05-23) <a href="https://doi.org/gc4jb3">https://doi.org/gc4jb3</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/s41559-017-0160">10.1038/s41559-017-0160</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/28812630">28812630</a></div></div>
</div>
<div id="ref-10ghgV3S8" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">12. </div><div class="csl-right-inline"><strong>Democratic databases: science on GitHub</strong> <div class="csl-block">Jeffrey Perkel</div> <em>Nature</em> (2016-10-03) <a href="https://doi.org/gdz6dq">https://doi.org/gdz6dq</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/538127a">10.1038/538127a</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/27708327">27708327</a></div></div>
</div>
<div id="ref-139b0pSGc" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">13. </div><div class="csl-right-inline"><strong>Social network of software development at GitHub</strong> <div class="csl-block">William Leibzon</div> <em>2016 IEEE/ACM International Conference on Advances in Social Networks Analysis and Mining (ASONAM)</em> (2016-08) <a href="https://doi.org/gph5qt">https://doi.org/gph5qt</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1109/asonam.2016.7752419">10.1109/asonam.2016.7752419</a></div></div>
</div>
<div id="ref-13QX8XU3J" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">14. </div><div class="csl-right-inline"><strong>A Beginner's Guide to Conducting Reproducible Research</strong> <div class="csl-block">Jesse M Alston, Jessica A Rick</div> <em>The Bulletin of the Ecological Society of America</em> (2021-01-15) <a href="https://doi.org/gk5v4p">https://doi.org/gk5v4p</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1002/bes2.1801">10.1002/bes2.1801</a></div></div>
</div>
<div id="ref-ydrk01SR" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">15. </div><div class="csl-right-inline"><strong>Foundational Practices of Research Data Management</strong> <div class="csl-block">Kristin Briney, Heather Coates, Abigail Goben</div> <em>Research Ideas and Outcomes</em> (2020-07-27) <a href="https://doi.org/ghssbk">https://doi.org/ghssbk</a> <div class="csl-block">DOI: <a href="https://doi.org/10.3897/rio.6.e56508">10.3897/rio.6.e56508</a></div></div>
</div>
<div id="ref-MXxgZJ45" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">16. </div><div class="csl-right-inline"><strong>Building tools with GitHub: customize your workflow</strong> <div class="csl-block">Chris Dawson</div> <em>O'Reilly</em> (2016) <div class="csl-block">ISBN: 9781491933503</div></div>
</div>
<div id="ref-1Du6fzB8g" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">17. </div><div class="csl-right-inline"><strong>A Guide to Using GitHub for Developing and Versioning Data Standards and Reporting Formats</strong> <div class="csl-block">Robert Crystal‐Ornelas, Charuleka Varadharajan, Ben Bond‐Lamberty, Kristin Boye, Madison Burrus, Shreyas Cholia, Michael Crow, Joan Damerow, Ranjeet Devarakonda, Kim S Ely, … Deborah A Agarwal</div> <em>Earth and Space Science</em> (2021-08) <a href="https://doi.org/gmbs9c">https://doi.org/gmbs9c</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1029/2021ea001797">10.1029/2021ea001797</a></div></div>
</div>
<div id="ref-uBJwnPbq" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">18. </div><div class="csl-right-inline"><strong>Elevating The Status of Code in Ecology</strong> <div class="csl-block">KAS Mislan, Jeffrey M Heer, Ethan P White</div> <em>Trends in Ecology &amp;amp; Evolution</em> (2016-01) <a href="https://doi.org/gg43mk">https://doi.org/gg43mk</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.tree.2015.11.006">10.1016/j.tree.2015.11.006</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26704455">26704455</a></div></div>
</div>
<div id="ref-WrDwZxcM" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">19. </div><div class="csl-right-inline"><strong>Barely sufficient practices in scientific computing</strong> <div class="csl-block">Graham Lee, Sebastian Bacon, Ian Bush, Laura Fortunato, David Gavaghan, Thibault Lestang, Caroline Morton, Martin Robinson, Philippe Rocca-Serra, Susanna-Assunta Sansone, Helena Webb</div> <em>Patterns</em> (2021-02) <a href="https://doi.org/gjpcb6">https://doi.org/gjpcb6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.patter.2021.100206">10.1016/j.patter.2021.100206</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/33659915">33659915</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7892476">PMC7892476</a></div></div>
</div>
<div id="ref-nxqXElqz" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">20. </div><div class="csl-right-inline"><strong>Open collaborative writing with Manubot</strong> <div class="csl-block">Daniel S Himmelstein, Vincent Rubinetti, David R Slochower, Dongbo Hu, Venkat S Malladi, Casey S Greene, Anthony Gitter</div> <em>PLOS Computational Biology</em> (2019-06-24) <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007128">https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007128</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pcbi.1007128">10.1371/journal.pcbi.1007128</a></div></div>
</div>
<div id="ref-1A97a4UwU" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">21. </div><div class="csl-right-inline"><strong>The Open Knowledge Foundation: Open Data Means Better Science</strong> <div class="csl-block">Jennifer C Molloy</div> <em>PLoS Biology</em> (2011-12-06) <a href="https://doi.org/g3b">https://doi.org/g3b</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1371/journal.pbio.1001195">10.1371/journal.pbio.1001195</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22162946">22162946</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3232214">PMC3232214</a></div></div>
</div>
<div id="ref-D4C4k4ak" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">22. </div><div class="csl-right-inline"><strong>Sustainable Research Software Hand-Over</strong> <div class="csl-block">J Fehr, C Himpe, S Rave, J Saak</div> <em>Journal of Open Research Software</em> (2021-04-30) <a href="https://doi.org/g4n4">https://doi.org/g4n4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.5334/jors.307">10.5334/jors.307</a></div></div>
</div>
<div id="ref-s91uGRZ2" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">23. </div><div class="csl-right-inline"><strong>About code owners</strong> <div class="csl-block">GitHub Docs</div> <a href="http://ghdocs-prod.azurewebsites.net:80/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners">http://ghdocs-prod.azurewebsites.net:80/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners</a></div>
</div>
<div id="ref-iIEKCTLU" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">24. </div><div class="csl-right-inline"><strong>The Tao of open science for ecology</strong> <div class="csl-block">Stephanie E Hampton, Sean S Anderson, Sarah C Bagby, Corinna Gries, Xueying Han, Edmund M Hart, Matthew B Jones, WChristopher Lenhardt, Andrew MacDonald, William K Michener, … Naupaka Zimmerman</div> <em>Ecosphere</em> (2015-07) <a href="https://doi.org/gdj5w6">https://doi.org/gdj5w6</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1890/es14-00402.1">10.1890/es14-00402.1</a></div></div>
</div>
<div id="ref-1GuRfvMeX" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">25. </div><div class="csl-right-inline"><strong>Ecological Data Should Not Be So Hard to Find and Reuse</strong> <div class="csl-block">Timothée Poisot, Anne Bruneau, Andrew Gonzalez, Dominique Gravel, Pedro Peres-Neto</div> <em>Trends in Ecology &amp;amp; Evolution</em> (2019-06) <a href="https://doi.org/gg43mw">https://doi.org/gg43mw</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1016/j.tree.2019.04.005">10.1016/j.tree.2019.04.005</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/31056219">31056219</a></div></div>
</div>
<div id="ref-6DjakjNS" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">26. </div><div class="csl-right-inline"><strong>The FAIR Guiding Principles for scientific data management and stewardship</strong> <div class="csl-block">Mark D Wilkinson, Michel Dumontier, IJsbrand Jan Aalbersberg, Gabrielle Appleton, Myles Axton, Arie Baak, Niklas Blomberg, Jan-Willem Boiten, Luiz Bonino da Silva Santos, Philip E Bourne, … Barend Mons</div> <em>Scientific Data</em> (2016-03-15) <a href="https://doi.org/bdd4">https://doi.org/bdd4</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/sdata.2016.18">10.1038/sdata.2016.18</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/26978244">26978244</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4792175">PMC4792175</a></div></div>
</div>
<div id="ref-Xsdcv6q" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">27. </div><div class="csl-right-inline"><strong>rfishbase: exploring, manipulating and visualizing FishBase data from R</strong> <div class="csl-block">C Boettiger, DT Lang, PC Wainwright</div> <em>Journal of Fish Biology</em> (2012-11) <a href="https://doi.org/gh5x27">https://doi.org/gh5x27</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1111/j.1095-8649.2012.03464.x">10.1111/j.1095-8649.2012.03464.x</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/23130696">23130696</a></div></div>
</div>
<div id="ref-FVBWKkZu" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">28. </div><div class="csl-right-inline"><strong>taxize: taxonomic search and retrieval in R</strong> <div class="csl-block">Scott A Chamberlain, Eduard Szöcs</div> <em>F1000Research</em> (2013-10-28) <a href="https://doi.org/ggdskx">https://doi.org/ggdskx</a> <div class="csl-block">DOI: <a href="https://doi.org/10.12688/f1000research.2-191.v2">10.12688/f1000research.2-191.v2</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/24555091">24555091</a> · PMCID: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3901538">PMC3901538</a></div></div>
</div>
<div id="ref-AE0QcVgJ" class="csl-entry" role="doc-biblioentry">
<div class="csl-left-margin">29. </div><div class="csl-right-inline"><strong>Synchronized editing: the future of collaborative writing</strong> <div class="csl-block">Jeffrey M Perkel</div> <em>Nature</em> (2020-03-31) <a href="https://doi.org/ggqk8s">https://doi.org/ggqk8s</a> <div class="csl-block">DOI: <a href="https://doi.org/10.1038/d41586-020-00916-6">10.1038/d41586-020-00916-6</a> · PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/32235940">32235940</a></div></div>
</div>
</div>
<!-- default theme -->

<style>
  /* import google fonts */
  @import url("https://fonts.googleapis.com/css?family=Open+Sans:400,600,700");
  @import url("https://fonts.googleapis.com/css?family=Source+Code+Pro");

  /* -------------------------------------------------- */
  /* global */
  /* -------------------------------------------------- */

  /* all elements */
  * {
    /* force sans-serif font unless specified otherwise */
    font-family: "Open Sans", "Helvetica", sans-serif;

    /* prevent text inflation on some mobile browsers */
    -webkit-text-size-adjust: none !important;
    -moz-text-size-adjust: none !important;
    -o-text-size-adjust: none !important;
    text-size-adjust: none !important;
  }

  @media only screen {
    /* "page" element */
    body {
      position: relative;
      box-sizing: border-box;
      font-size: 12pt;
      line-height: 1.5;
      max-width: 8.5in;
      margin: 20px auto;
      padding: 40px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* "page" element */
    body {
      padding: 20px;
      margin: 0;
      border-radius: 0;
      border: none;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05) inset;
      background: none;
    }
  }

  /* -------------------------------------------------- */
  /* headings */
  /* -------------------------------------------------- */

  /* all headings */
  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    margin: 20px 0;
    padding: 0;
    font-weight: bold;
  }

  /* biggest heading */
  h1 {
    margin: 40px 0;
    text-align: center;
  }

  /* second biggest heading */
  h2 {
    margin-top: 30px;
    padding-bottom: 5px;
    border-bottom: solid 1px #bdbdbd;
  }

  /* heading font sizes */
  h1 {
    font-size: 2em;
  }
  h2 {
    font-size: 1.5em;
  }
  h3 {
    font-size: 1.35em;
  }
  h4 {
    font-size: 1.25em;
  }
  h5 {
    font-size: 1.15em;
  }
  h6 {
    font-size: 1em;
  }

  /* -------------------------------------------------- */
  /* manuscript header */
  /* -------------------------------------------------- */

  /* manuscript title */
  header > h1 {
    margin: 0;
  }

  /* manuscript title caption text (ie "automatically generated on") */
  header + p {
    text-align: center;
    margin-top: 10px;
  }

  /* -------------------------------------------------- */
  /* text elements */
  /* -------------------------------------------------- */

  /* links */
  a {
    color: #2196f3;
    overflow-wrap: break-word;
  }

  /* superscripts and subscripts */
  sub,
  sup {
    /* prevent from affecting line height */
    line-height: 0;
  }

  /* unordered and ordered lists*/
  ul,
  ol {
    padding-left: 20px;
  }

  /* class for styling text semibold */
  .semibold {
    font-weight: 600;
  }

  /* class for styling elements horizontally left aligned */
  .left {
    display: block;
    text-align: left;
    margin-left: auto;
    margin-right: 0;
    justify-content: left;
  }

  /* class for styling elements horizontally centered */
  .center {
    display: block;
    text-align: center;
    margin-left: auto;
    margin-right: auto;
    justify-content: center;
  }

  /* class for styling elements horizontally right aligned */
  .right {
    display: block;
    text-align: right;
    margin-left: 0;
    margin-right: auto;
    justify-content: right;
  }

  /* -------------------------------------------------- */
  /* section elements */
  /* -------------------------------------------------- */

  /* horizontal divider line */
  hr {
    border: none;
    height: 1px;
    background: #bdbdbd;
  }

  /* paragraphs, horizontal dividers, figures, tables, code */
  p,
  hr,
  figure,
  table,
  pre {
    /* treat all as "paragraphs", with consistent vertical margins */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* figures */
  /* -------------------------------------------------- */

  /* figure */
  figure {
    max-width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure caption */
  figcaption {
    padding: 0;
    padding-top: 10px;
  }

  /* figure image element */
  figure > img,
  figure > svg {
    max-width: 100%;
    display: block;
    margin-left: auto;
    margin-right: auto;
  }

  /* figure auto-number */
  img + figcaption > span:first-of-type,
  svg + figcaption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* tables */
  /* -------------------------------------------------- */

  /* table */
  table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
  }

  /* table cells */
  th,
  td {
    border: solid 1px #bdbdbd;
    padding: 10px;
    /* squash table if too wide for page by forcing line breaks */
    overflow-wrap: break-word;
    word-break: break-word;
  }

  /* header row and even rows */
  th,
  tr:nth-child(2n) {
    background-color: #fafafa;
  }

  /* odd rows */
  tr:nth-child(2n + 1) {
    background-color: #ffffff;
  }

  /* table caption */
  caption {
    text-align: left;
    padding: 0;
    padding-bottom: 10px;
  }

  /* table auto-number */
  table > caption > span:first-of-type {
    font-weight: bold;
    margin-right: 5px;
  }

  /* -------------------------------------------------- */
  /* code */
  /* -------------------------------------------------- */

  /* multi-line code block */
  pre {
    padding: 10px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
    break-inside: avoid;
    text-align: left;
  }

  /* inline code, ie code within normal text */
  :not(pre) > code {
    padding: 0 4px;
    background-color: #eeeeee;
    color: #000000;
    border-radius: 5px;
  }

  /* code text */
  /* apply all children, to reach syntax highlighting sub-elements */
  code,
  code * {
    /* force monospace font */
    font-family: "Source Code Pro", "Courier New", monospace;
  }

  /* -------------------------------------------------- */
  /* quotes */
  /* -------------------------------------------------- */

  /* quoted text */
  blockquote {
    margin: 0;
    padding: 0;
    border-left: 4px solid #bdbdbd;
    padding-left: 16px;
    break-inside: avoid;
  }

  /* -------------------------------------------------- */
  /* banners */
  /* -------------------------------------------------- */

  /* info banners */
  .banner {
    box-sizing: border-box;
    display: block;
    position: relative;
    width: 100%;
    margin-top: 20px;
    margin-bottom: 20px;
    padding: 20px;
    text-align: center;
  }

  /* paragraph in banner */
  .banner > p {
    margin: 0;
  }

  /* -------------------------------------------------- */
  /* highlight colors */
  /* -------------------------------------------------- */

  .white {
    background: #ffffff;
  }
  .lightgrey {
    background: #eeeeee;
  }
  .grey {
    background: #757575;
  }
  .darkgrey {
    background: #424242;
  }
  .black {
    background: #000000;
  }
  .lightred {
    background: #ffcdd2;
  }
  .lightyellow {
    background: #ffecb3;
  }
  .lightgreen {
    background: #dcedc8;
  }
  .lightblue {
    background: #e3f2fd;
  }
  .lightpurple {
    background: #f3e5f5;
  }
  .red {
    background: #f44336;
  }
  .orange {
    background: #ff9800;
  }
  .yellow {
    background: #ffeb3b;
  }
  .green {
    background: #4caf50;
  }
  .blue {
    background: #2196f3;
  }
  .purple {
    background: #9c27b0;
  }
  .white,
  .lightgrey,
  .lightred,
  .lightyellow,
  .lightgreen,
  .lightblue,
  .lightpurple,
  .orange,
  .yellow,
  .white a,
  .lightgrey a,
  .lightred a,
  .lightyellow a,
  .lightgreen a,
  .lightblue a,
  .lightpurple a,
  .orange a,
  .yellow a {
    color: #000000;
  }
  .grey,
  .darkgrey,
  .black,
  .red,
  .green,
  .blue,
  .purple,
  .grey a,
  .darkgrey a,
  .black a,
  .red a,
  .green a,
  .blue a,
  .purple a {
    color: #ffffff;
  }

  /* -------------------------------------------------- */
  /* buttons */
  /* -------------------------------------------------- */

  /* class for styling links like buttons */
  .button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin: 5px;
    padding: 10px 20px;
    font-size: 0.75em;
    font-weight: 600;
    text-transform: uppercase;
    text-decoration: none;
    letter-spacing: 1px;
    background: none;
    color: #2196f3;
    border: solid 1px #bdbdbd;
    border-radius: 5px;
  }

  /* buttons when hovered */
  .button:hover:not([disabled]),
  .icon_button:hover:not([disabled]) {
    cursor: pointer;
    background: #f5f5f5;
  }

  /* buttons when disabled */
  .button[disabled],
  .icon_button[disabled] {
    opacity: 0.35;
    pointer-events: none;
  }

  /* class for styling buttons containg only single icon */
  .icon_button {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    text-decoration: none;
    margin: 0;
    padding: 0;
    background: none;
    border-radius: 5px;
    border: none;
    width: 20px;
    height: 20px;
    min-width: 20px;
    min-height: 20px;
  }

  /* icon button inner svg image */
  .icon_button > svg {
    height: 16px;
  }

  /* -------------------------------------------------- */
  /* icons */
  /* -------------------------------------------------- */

  /* class for styling icons inline with text */
  .inline_icon {
    height: 1em;
    position: relative;
    top: 0.125em;
  }

  /* -------------------------------------------------- */
  /* references */
  /* -------------------------------------------------- */

  .csl-entry {
    margin-top: 15px;
    margin-bottom: 15px;
  }

  /* -------------------------------------------------- */
  /* print control */
  /* -------------------------------------------------- */

  @media print {
    @page {
      /* suggested printing margin */
      margin: 0.5in;
    }

    /* document and "page" elements */
    html,
    body {
      margin: 0;
      padding: 0;
      width: 100%;
      height: 100%;
    }

    /* "page" element */
    body {
      font-size: 11pt !important;
      line-height: 1.35;
    }

    /* all headings */
    h1,
    h2,
    h3,
    h4,
    h5,
    h6 {
      margin: 15px 0;
    }

    /* figures and tables */
    figure,
    table {
      font-size: 0.85em;
    }

    /* table cells */
    th,
    td {
      padding: 5px;
    }

    /* shrink font awesome icons */
    i.fas,
    i.fab,
    i.far,
    i.fal {
      transform: scale(0.85);
    }

    /* decrease banner margins */
    .banner {
      margin-top: 15px;
      margin-bottom: 15px;
      padding: 15px;
    }

    /* class for centering an element vertically on its own page */
    .page_center {
      margin: auto;
      width: 100%;
      height: 100%;
      display: flex;
      align-items: center;
      vertical-align: middle;
      break-before: page;
      break-after: page;
    }

    /* always insert a page break before the element */
    .page_break_before {
      break-before: page;
    }

    /* always insert a page break after the element */
    .page_break_after {
      break-after: page;
    }

    /* avoid page break before the element */
    .page_break_before_avoid {
      break-before: avoid;
    }

    /* avoid page break after the element */
    .page_break_after_avoid {
      break-after: avoid;
    }

    /* avoid page break inside the element */
    .page_break_inside_avoid {
      break-inside: avoid;
    }
  }

  /* -------------------------------------------------- */
  /* override pandoc css quirks */
  /* -------------------------------------------------- */

  .sourceCode {
    /* prevent unsightly overflow in wide code blocks */
    overflow: auto !important;
  }

  div.sourceCode {
    /* prevent background fill on top-most code block  container */
    background: none !important;
  }

  .sourceCode * {
    /* force consistent line spacing */
    line-height: 1.5 !important;
  }

  div.sourceCode {
    /* style code block margins same as <pre> element */
    margin-top: 20px;
    margin-bottom: 20px;
  }

  /* -------------------------------------------------- */
  /* tablenos */
  /* -------------------------------------------------- */

  /* tablenos wrapper */
  .tablenos {
    width: 100%;
    margin: 20px 0;
  }

  .tablenos > table {
    /* move margins from table to table_wrapper to allow margin collapsing */
    margin: 0;
  }

  @media only screen {
    /* tablenos wrapper */
    .tablenos {
      /* show scrollbar on tables if necessary to prevent overflow */
      overflow-x: auto !important;
    }

    .tablenos th,
    .tablenos td {
      overflow-wrap: unset !important;
      word-break: unset !important;
    }

    /* table in wrapper */
    .tablenos table,
    .tablenos table * {
      /* don't break table words */
      overflow-wrap: normal !important;
    }
  }
</style>
<!-- 
    Plugin Core

    Functions needed for and shared across all first-party plugins.
-->

<script>
  // get element that is target of hash (from link element or url)
  function getHashTarget(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector(`[id="${id}"]`);
    if (!target) return;

    // if figure or table, modify target to get expected element
    if (id.indexOf("fig:") === 0) target = target.querySelector("figure");
    if (id.indexOf("tbl:") === 0) target = target.querySelector("table");

    return target;
  }

  // get position/dimensions of element or viewport
  function getRectInView(element) {
    let rect = {};
    rect.left = 0;
    rect.top = 0;
    rect.right = document.documentElement.clientWidth;
    rect.bottom = document.documentElement.clientHeight;
    let style = {};

    if (element instanceof HTMLElement) {
      rect = element.getBoundingClientRect();
      style = window.getComputedStyle(element);
    }

    const margin = {};
    margin.left = parseFloat(style.marginLeftWidth) || 0;
    margin.top = parseFloat(style.marginTopWidth) || 0;
    margin.right = parseFloat(style.marginRightWidth) || 0;
    margin.bottom = parseFloat(style.marginBottomWidth) || 0;

    const border = {};
    border.left = parseFloat(style.borderLeftWidth) || 0;
    border.top = parseFloat(style.borderTopWidth) || 0;
    border.right = parseFloat(style.borderRightWidth) || 0;
    border.bottom = parseFloat(style.borderBottomWidth) || 0;

    const newRect = {};
    newRect.left = rect.left + margin.left + border.left;
    newRect.top = rect.top + margin.top + border.top;
    newRect.right = rect.right + margin.right + border.right;
    newRect.bottom = rect.bottom + margin.bottom + border.bottom;
    newRect.width = newRect.right - newRect.left;
    newRect.height = newRect.bottom - newRect.top;

    return newRect;
  }

  // get position of element relative to page
  function getRectInPage(element) {
    const rect = getRectInView(element);
    const body = getRectInView(document.body);

    const newRect = {};
    newRect.left = rect.left - body.left;
    newRect.top = rect.top - body.top;
    newRect.right = rect.right - body.left;
    newRect.bottom = rect.bottom - body.top;
    newRect.width = rect.width;
    newRect.height = rect.height;

    return newRect;
  }

  // get closest element before specified element that matches query
  function firstBefore(element, query) {
    while (element && element !== document.body && !element.matches(query))
      element = element.previousElementSibling || element.parentNode;

    return element;
  }

  // check if element is part of collapsed heading
  function isCollapsed(element) {
    while (element && element !== document.body) {
      if (element.dataset.collapsed === "true") return true;
      element = element.parentNode;
    }
    return false;
  }

  // expand any collapsed parent containers of element if necessary
  function expandElement(element) {
    if (isCollapsed(element)) {
      // accordion plugin
      const heading = firstBefore(element, "h2");
      if (heading) heading.click();
      // details/summary HTML element
      const summary = firstBefore(element, "summary");
      if (summary) summary.click();
    }
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);

    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // get list of elements after a start element up to element matching query
  function nextUntil(element, query, exclude) {
    const elements = [];
    while (((element = element.nextElementSibling), element)) {
      if (element.matches(query)) break;
      if (!element.matches(exclude)) elements.push(element);
    }
    return elements;
  }
</script>
<!--
  Accordion Plugin

  Allows sections of content under h2 headings to be collapsible.
-->

<script type="module">
  // whether to always start expanded ('false'), always start collapsed
  // ('true'), or start collapsed when screen small ('auto')
  const startCollapsed = "auto";

  // start script
  function start() {
    // run through each <h2> heading
    const headings = document.querySelectorAll("h2");
    for (const heading of headings) {
      addArrow(heading);

      // start expanded/collapsed based on option
      if (
        startCollapsed === "true" ||
        (startCollapsed === "auto" && isSmallScreen()) ||
        heading.dataset.collapsed === "true"
      )
        collapseHeading(heading);
      else expandElement(heading);
    }

    // attach hash change listener to window
    window.addEventListener("hashchange", onHashChange);
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) goToElement(target);
  }

  // add arrow to heading
  function addArrow(heading) {
    // add arrow button
    const arrow = document.createElement("button");
    arrow.innerHTML = document.querySelector(".icon_angle_down").innerHTML;
    arrow.classList.add("icon_button", "accordion_arrow");
    heading.insertBefore(arrow, heading.firstChild);

    // attach click listener to heading and button
    heading.addEventListener("click", onHeadingClick);
    arrow.addEventListener("click", onArrowClick);
  }

  // determine if on mobile-like device with small screen
  function isSmallScreen() {
    return Math.min(window.innerWidth, window.innerHeight) < 480;
  }

  // when <h2> heading is clicked
  function onHeadingClick(event) {
    // only collapse if <h2> itself is target of click (eg, user did
    // not click on anchor within <h2>)
    if (event.target === this) toggleCollapse(this);
  }

  // when arrow button is clicked
  function onArrowClick() {
    toggleCollapse(this.parentNode);
  }

  // collapse section if expanded, expand if collapsed
  function toggleCollapse(heading) {
    if (heading.dataset.collapsed === "false") collapseHeading(heading);
    else expandElement(heading);
  }

  // elements to exclude from collapse, such as table of contents panel,
  // hypothesis panel, etc
  const exclude = "#toc_panel, div.annotator-frame, #lightbox_overlay";

  // collapse section
  function collapseHeading(heading) {
    heading.setAttribute("data-collapsed", "true");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "true");
  }

  // expand section
  function expandElement(heading) {
    heading.setAttribute("data-collapsed", "false");
    const children = getChildren(heading);
    for (const child of children) child.setAttribute("data-collapsed", "false");
  }

  // get list of elements between this <h2> and next <h2> or <h1>
  // ("children" of the <h2> section)
  function getChildren(heading) {
    return nextUntil(heading, "h2, h1", exclude);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle down icon -->

<template class="icon_angle_down">
  <!-- modified from: https://fontawesome.com/icons/angle-down -->
  <svg width="16" height="16" viewBox="0 0 448 512">
    <path
      fill="currentColor"
      d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* accordion arrow button */
    .accordion_arrow {
      margin-right: 10px;
    }

    /* arrow icon when <h2> data-collapsed attribute true */
    h2[data-collapsed="true"] > .accordion_arrow > svg {
      transform: rotate(-90deg);
    }

    /* all elements (except <h2>'s) when data-collapsed attribute true */
    *:not(h2)[data-collapsed="true"] {
      display: none;
    }

    /* accordion arrow button when hovered and <h2>'s when hovered */
    .accordion_arrow:hover,
    h2[data-collapsed="true"]:hover,
    h2[data-collapsed="false"]:hover {
      cursor: pointer;
    }
  }

  /* always hide accordion arrow button on print */
  @media only print {
    .accordion_arrow {
      display: none;
    }
  }
</style>
<!--
  Anchors Plugin

  Adds an anchor next to each of a certain type of element that provides a
  human-readable url to that specific item/position in the document (e.g.
  "manuscript.html#abstract"). It also makes it such that scrolling out of view
  of a target removes its identifier from the url.
-->

<script type="module">
  // which types of elements to add anchors next to, in "document.querySelector"
  // format
  const typesQuery =
    'h1, h2, h3, div[id^="fig:"], div[id^="tbl:"], span[id^="eq:"]';

  // start script
  function start() {
    // add anchor to each element of specified types
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) addAnchor(element);

    // attach scroll listener to window
    window.addEventListener("scroll", onScroll);
  }

  // when window is scrolled
  function onScroll() {
    // if url has hash and user has scrolled out of view of hash
    // target, remove hash from url
    const tolerance = 100;
    const target = getHashTarget();
    if (target) {
      if (
        target.getBoundingClientRect().top > window.innerHeight + tolerance ||
        target.getBoundingClientRect().bottom < 0 - tolerance
      )
        history.pushState(null, null, " ");
    }
  }

  // add anchor to element
  function addAnchor(element) {
    let addTo; // element to add anchor button to

    // if figure or table, modify withId and addTo to get expected
    // elements
    if (element.id.indexOf("fig:") === 0) {
      addTo = element.querySelector("figcaption");
    } else if (element.id.indexOf("tbl:") === 0) {
      addTo = element.querySelector("caption");
    } else if (element.id.indexOf("eq:") === 0) {
      addTo = element.querySelector(".eqnos-number");
    }

    addTo = addTo || element;
    const id = element.id || null;

    // do not add anchor if element doesn't have assigned id.
    // id is generated by pandoc and is assumed to be unique and
    // human-readable
    if (!id) return;

    // create anchor button
    const anchor = document.createElement("a");
    anchor.innerHTML = document.querySelector(".icon_link").innerHTML;
    anchor.title = "Link to this part of the document";
    anchor.classList.add("icon_button", "anchor");
    anchor.dataset.ignore = "true";
    anchor.href = "#" + id;
    addTo.appendChild(anchor);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- link icon -->

<template class="icon_link">
  <!-- modified from: https://fontawesome.com/icons/link -->
  <svg width="16" height="16" viewBox="0 0 512 512">
    <path
      fill="currentColor"
      d="M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* anchor button */
    .anchor {
      opacity: 0;
      margin-left: 5px;
    }

    /* anchor buttons within <h2>'s */
    h2 .anchor {
      margin-left: 10px;
    }

    /* anchor buttons when hovered/focused and anything containing an anchor button when hovered */
    *:hover > .anchor,
    .anchor:hover,
    .anchor:focus {
      opacity: 1;
    }

    /* anchor button when hovered */
    .anchor:hover {
      cursor: pointer;
    }
  }

  /* always show anchor button on devices with no mouse/hover ability */
  @media (hover: none) {
    .anchor {
      opacity: 1;
    }
  }

  /* always hide anchor button on print */
  @media only print {
    .anchor {
      display: none;
    }
  }
</style>
<!-- 
    Attributes Plugin

    Allows arbitrary HTML attributes to be attached to (almost) any element.
    Place an HTML comment inside or next to the desired element with the content:
    $attribute="value"
-->

<script type="module">
  // start script
  function start() {
    // get list of comments in document
    const comments = findComments();

    for (const comment of comments)
      if (comment.parentElement)
        addAttributes(comment.parentElement, comment.nodeValue.trim());
  }

  // add html attributes to specified element based on string of
  // html attributes and values
  function addAttributes(element, text) {
    // regex's for finding attribute/value pairs in the format of
    // attribute="value" or attribute='value
    const regex2 = /\$([a-zA-Z\-]+)?=\"(.+?)\"/;
    const regex1 = /\$([a-zA-Z\-]+)?=\'(.+?)\'/;

    // loop through attribute/value pairs
    let match;
    while ((match = text.match(regex2) || text.match(regex1))) {
      // get attribute and value from regex capture groups
      let attribute = match[1];
      let value = match[2];

      // remove from string
      text = text.substring(match.index + match[0].length);

      if (!attribute || !value) break;

      // set attribute of parent element
      try {
        element.setAttribute(attribute, value);
      } catch (error) {
        console.log(error);
      }

      // special case for colspan
      if (attribute === "colspan") removeTableCells(element, value);
    }
  }

  // get list of comment elements in document
  function findComments() {
    const comments = [];

    // iterate over comment nodes in document
    function acceptNode(node) {
      return NodeFilter.FILTER_ACCEPT;
    }
    const iterator = document.createNodeIterator(
      document.body,
      NodeFilter.SHOW_COMMENT,
      acceptNode
    );
    let node;
    while ((node = iterator.nextNode())) comments.push(node);

    return comments;
  }

  // remove certain number of cells after specified cell
  function removeTableCells(cell, number) {
    number = parseInt(number);
    if (!number) return;

    // remove elements
    for (; number > 1; number--) {
      if (cell.nextElementSibling) cell.nextElementSibling.remove();
    }
  }

  // start script on DOMContentLoaded instead of load to ensure this plugins
  // runs before other plugins
  window.addEventListener("DOMContentLoaded", start);
</script>
<!--
  Jump to First Plugin

  Adds a button next to each reference entry, figure, and table that jumps the
  page to the first occurrence of a link to that item in the manuscript.
-->

<script type="module">
  // whether to add buttons next to reference entries
  const references = "true";
  // whether to add buttons next to figures
  const figures = "true";
  // whether to add buttons next to tables
  const tables = "true";

  // start script
  function start() {
    if (references !== "false")
      makeButtons(`div[id^="ref-"]`, ".csl-left-margin", "reference");
    if (figures !== "false")
      makeButtons(`div[id^="fig:"]`, "figcaption", "figure");
    if (tables !== "false") makeButtons(`div[id^="tbl:"]`, "caption", "table");
  }

  // when jump button clicked
  function onButtonClick() {
    const first = getFirstOccurrence(this.dataset.id);
    if (!first) return;

    // update url hash so navigating "back" in history will return user to button
    window.location.hash = this.dataset.id;
    // scroll to link
    const timeout = function () {
      goToElement(first, window.innerHeight * 0.5);
    };
    window.setTimeout(timeout, 0);
  }

  // get first occurrence of link to item in document
  function getFirstOccurrence(id) {
    let query = "a";
    query += '[href="#' + id + '"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelector(query);
  }

  // add button next to each reference entry, figure, or table
  function makeButtons(query, containerQuery, subject) {
    const elements = document.querySelectorAll(query);
    for (const element of elements) {
      const id = element.id;
      const buttonContainer = element.querySelector(containerQuery);
      const first = getFirstOccurrence(id);

      // if can't find link to reference or place to put button, ignore
      if (!first || !buttonContainer) continue;

      // make jump button
      let button = document.createElement("button");
      button.classList.add("icon_button", "jump_arrow");
      button.title = `Jump to the first occurrence of this ${subject} in the document`;
      const icon = document.querySelector(".icon_angle_double_up");
      button.innerHTML = icon.innerHTML;
      button.dataset.id = id;
      button.dataset.ignore = "true";
      button.addEventListener("click", onButtonClick);
      buttonContainer.prepend(button);
    }
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- angle double up icon -->

<template class="icon_angle_double_up">
  <!-- modified from: https://fontawesome.com/icons/angle-double-up -->
  <svg width="16" height="16" viewBox="0 0 320 512">
    <path
      fill="currentColor"
      d="M177 255.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 351.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 425.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1zm-34-192L7 199.7c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l96.4-96.4 96.4 96.4c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9l-136-136c-9.2-9.4-24.4-9.4-33.8 0z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* jump button */
    .jump_arrow {
      position: relative;
      top: 0.125em;
      margin-right: 5px;
    }
  }

  /* always hide jump button on print */
  @media only print {
    .jump_arrow {
      display: none;
    }
  }
</style>
<!-- 
    Lightbox Plugin

    Makes it such that when a user clicks on an image, the image fills the
    screen and the user can pan/drag/zoom the image and navigate between other
    images in the document.
-->

<script type="module">
  // list of possible zoom/scale factors
  const zooms =
    "0.1, 0.25, 0.333333, 0.5, 0.666666, 0.75, 1, 1.25, 1.5, 1.75, 2, 2.5, 3, 3.5, 4, 5, 6, 7, 8";
  // whether to fit image to view ('fit'), display at 100% and shrink if
  // necessary ('shrink'), or always display at 100% ('100')
  const defaultZoom = "fit";
  // whether to zoom in/out toward center of view ('true') or mouse ('false')
  const centerZoom = "false";

  // start script
  function start() {
    // run through each <img> element
    const imgs = document.querySelectorAll("figure > img");
    let count = 1;
    for (const img of imgs) {
      img.classList.add("lightbox_document_img");
      img.dataset.number = count;
      img.dataset.total = imgs.length;
      img.addEventListener("click", openLightbox);
      count++;
    }

    // attach mouse and key listeners to window
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("keyup", onKeyUp);
  }

  // when mouse is moved anywhere in window
  function onWindowMouseMove(event) {
    window.mouseX = event.clientX;
    window.mouseY = event.clientY;
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("lightbox_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("lightbox_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeLightbox();
        break;
    }
  }

  // open lightbox
  function openLightbox() {
    const lightbox = makeLightbox(this);
    if (!lightbox) return;

    blurBody(lightbox);
    document.body.appendChild(lightbox);
  }

  // make lightbox
  function makeLightbox(img) {
    // delete lightbox if it exists, start fresh
    closeLightbox();

    // create screen overlay containing lightbox
    const overlay = document.createElement("div");
    overlay.id = "lightbox_overlay";

    // create image info boxes
    const numberInfo = document.createElement("div");
    const zoomInfo = document.createElement("div");
    numberInfo.id = "lightbox_number_info";
    zoomInfo.id = "lightbox_zoom_info";

    // create container for image
    const imageContainer = document.createElement("div");
    imageContainer.id = "lightbox_image_container";
    const lightboxImg = makeLightboxImg(
      img,
      imageContainer,
      numberInfo,
      zoomInfo
    );
    imageContainer.appendChild(lightboxImg);

    // create bottom container for caption and navigation buttons
    const bottomContainer = document.createElement("div");
    bottomContainer.id = "lightbox_bottom_container";
    const caption = makeCaption(img);
    const prevButton = makePrevButton(img);
    const nextButton = makeNextButton(img);
    bottomContainer.appendChild(prevButton);
    bottomContainer.appendChild(caption);
    bottomContainer.appendChild(nextButton);

    // attach top middle and bottom to overlay
    overlay.appendChild(numberInfo);
    overlay.appendChild(zoomInfo);
    overlay.appendChild(imageContainer);
    overlay.appendChild(bottomContainer);

    return overlay;
  }

  // make <img> object that is intuitively draggable and zoomable
  function makeLightboxImg(sourceImg, container, numberInfoBox, zoomInfoBox) {
    // create copy of source <img>
    const img = sourceImg.cloneNode(true);
    img.classList.remove("lightbox_document_img");
    img.removeAttribute("id");
    img.removeAttribute("width");
    img.removeAttribute("height");
    img.style.position = "unset";
    img.style.margin = "0";
    img.style.padding = "0";
    img.style.width = "";
    img.style.height = "";
    img.style.minWidth = "";
    img.style.minHeight = "";
    img.style.maxWidth = "";
    img.style.maxHeight = "";
    img.id = "lightbox_img";

    // build sorted list of zoomSteps
    const zoomSteps = zooms.split(/[^0-9.]/).map((step) => parseFloat(step));
    zoomSteps.sort((a, b) => a - b);

    // <img> object property variables
    let zoom = 1;
    let translateX = 0;
    let translateY = 0;
    let clickMouseX = undefined;
    let clickMouseY = undefined;
    let clickTranslateX = undefined;
    let clickTranslateY = undefined;

    updateNumberInfo();

    // update image numbers displayed in info box
    function updateNumberInfo() {
      numberInfoBox.innerHTML =
        sourceImg.dataset.number + " of " + sourceImg.dataset.total;
    }

    // update zoom displayed in info box
    function updateZoomInfo() {
      let zoomInfo = zoom * 100;
      if (!Number.isInteger(zoomInfo)) zoomInfo = zoomInfo.toFixed(2);
      zoomInfoBox.innerHTML = zoomInfo + "%";
    }

    // move to closest zoom step above current zoom
    const zoomIn = function () {
      for (const zoomStep of zoomSteps) {
        if (zoomStep > zoom) {
          zoom = zoomStep;
          break;
        }
      }
      updateTransform();
    };

    // move to closest zoom step above current zoom
    const zoomOut = function () {
      zoomSteps.reverse();
      for (const zoomStep of zoomSteps) {
        if (zoomStep < zoom) {
          zoom = zoomStep;
          break;
        }
      }
      zoomSteps.reverse();

      updateTransform();
    };

    // update display of <img> based on scale/translate properties
    const updateTransform = function () {
      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      // get new width/height after scale
      const rect = img.getBoundingClientRect();
      // limit translate
      translateX = Math.max(translateX, -rect.width / 2);
      translateX = Math.min(translateX, rect.width / 2);
      translateY = Math.max(translateY, -rect.height / 2);
      translateY = Math.min(translateY, rect.height / 2);

      // set transform
      img.style.transform =
        "translate(" +
        (translateX || 0) +
        "px," +
        (translateY || 0) +
        "px) scale(" +
        (zoom || 1) +
        ")";

      updateZoomInfo();
    };

    // fit <img> to container
    const fit = function () {
      // no x/y offset, 100% zoom by default
      translateX = 0;
      translateY = 0;
      zoom = 1;

      // widths of <img> and container
      const imgWidth = img.naturalWidth;
      const imgHeight = img.naturalHeight;
      const containerWidth = parseFloat(
        window.getComputedStyle(container).width
      );
      const containerHeight = parseFloat(
        window.getComputedStyle(container).height
      );

      // how much zooming is needed to fit <img> to container
      const xRatio = imgWidth / containerWidth;
      const yRatio = imgHeight / containerHeight;
      const maxRatio = Math.max(xRatio, yRatio);
      const newZoom = 1 / maxRatio;

      // fit <img> to container according to option
      if (defaultZoom === "shrink") {
        if (maxRatio > 1) zoom = newZoom;
      } else if (defaultZoom === "fit") zoom = newZoom;

      updateTransform();
    };

    // when mouse wheel is rolled anywhere in container
    const onContainerWheel = function (event) {
      if (!event) return;

      // let ctrl + mouse wheel to zoom behave as normal
      if (event.ctrlKey) return;

      // prevent normal scroll behavior
      event.preventDefault();
      event.stopPropagation();

      // point around which to scale img
      const viewRect = container.getBoundingClientRect();
      const viewX = (viewRect.left + viewRect.right) / 2;
      const viewY = (viewRect.top + viewRect.bottom) / 2;
      const originX = centerZoom === "true" ? viewX : mouseX;
      const originY = centerZoom === "true" ? viewY : mouseY;

      // get point on image under origin
      const oldRect = img.getBoundingClientRect();
      const oldPercentX = (originX - oldRect.left) / oldRect.width;
      const oldPercentY = (originY - oldRect.top) / oldRect.height;

      // increment/decrement zoom
      if (event.deltaY < 0) zoomIn();
      if (event.deltaY > 0) zoomOut();

      // get offset between previous image point and origin
      const newRect = img.getBoundingClientRect();
      const offsetX = originX - (newRect.left + newRect.width * oldPercentX);
      const offsetY = originY - (newRect.top + newRect.height * oldPercentY);

      // translate image to keep image point under origin
      translateX += offsetX;
      translateY += offsetY;

      // perform translate
      updateTransform();
    };

    // when container is clicked
    function onContainerClick(event) {
      // if container itself is target of click, and not other
      // element above it
      if (event.target === this) closeLightbox();
    }

    // when mouse button is pressed on image
    const onImageMouseDown = function (event) {
      // store original mouse position relative to image
      clickMouseX = window.mouseX;
      clickMouseY = window.mouseY;
      clickTranslateX = translateX;
      clickTranslateY = translateY;
      event.stopPropagation();
      event.preventDefault();
    };

    // when mouse button is released anywhere in window
    const onWindowMouseUp = function (event) {
      // reset original mouse position
      clickMouseX = undefined;
      clickMouseY = undefined;
      clickTranslateX = undefined;
      clickTranslateY = undefined;

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mouseup", onWindowMouseUp);
    };

    // when mouse is moved anywhere in window
    const onWindowMouseMove = function (event) {
      if (
        clickMouseX === undefined ||
        clickMouseY === undefined ||
        clickTranslateX === undefined ||
        clickTranslateY === undefined
      )
        return;

      // offset image based on original and current mouse position
      translateX = clickTranslateX + window.mouseX - clickMouseX;
      translateY = clickTranslateY + window.mouseY - clickMouseY;
      updateTransform();
      event.preventDefault();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("mousemove", onWindowMouseMove);
    };

    // when window is resized
    const onWindowResize = function (event) {
      fit();

      // remove global listener if lightbox removed from document
      if (!document.body.contains(container))
        window.removeEventListener("resize", onWindowResize);
    };

    // attach the necessary event listeners
    img.addEventListener("dblclick", fit);
    img.addEventListener("mousedown", onImageMouseDown);
    container.addEventListener("wheel", onContainerWheel);
    container.addEventListener("mousedown", onContainerClick);
    container.addEventListener("touchstart", onContainerClick);
    window.addEventListener("mouseup", onWindowMouseUp);
    window.addEventListener("mousemove", onWindowMouseMove);
    window.addEventListener("resize", onWindowResize);

    // run fit() after lightbox atttached to document and <img> Loaded
    // so needed container and img dimensions available
    img.addEventListener("load", fit);

    return img;
  }

  // make caption
  function makeCaption(img) {
    const caption = document.createElement("div");
    caption.id = "lightbox_caption";
    const captionSource = img.nextElementSibling;
    if (captionSource.tagName.toLowerCase() === "figcaption") {
      const captionCopy = makeCopy(captionSource);
      caption.innerHTML = captionCopy.innerHTML;
    }

    caption.addEventListener("touchstart", function (event) {
      event.stopPropagation();
    });

    return caption;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // make button to jump to previous image in document
  function makePrevButton(img) {
    const prevButton = document.createElement("button");
    prevButton.id = "lightbox_prev_button";
    prevButton.title = "Jump to the previous image in the document [←]";
    prevButton.classList.add("icon_button", "lightbox_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;

    // attach click listeners to button
    prevButton.addEventListener("click", function () {
      getPrevImg(img).click();
    });

    return prevButton;
  }

  // make button to jump to next image in document
  function makeNextButton(img) {
    const nextButton = document.createElement("button");
    nextButton.id = "lightbox_next_button";
    nextButton.title = "Jump to the next image in the document [→]";
    nextButton.classList.add("icon_button", "lightbox_button");
    nextButton.innerHTML = document.querySelector(
      ".icon_caret_right"
    ).innerHTML;

    // attach click listeners to button
    nextButton.addEventListener("click", function () {
      getNextImg(img).click();
    });

    return nextButton;
  }

  // get previous image in document
  function getPrevImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if < 1
    if (index - 1 >= 0) index--;
    else index = imgs.length - 1;
    return imgs[index];
  }

  // get next image in document
  function getNextImg(img) {
    const imgs = document.querySelectorAll(".lightbox_document_img");

    // find index of provided img
    let index;
    for (index = 0; index < imgs.length; index++) {
      if (imgs[index] === img) break;
    }

    // wrap index to other side if > total
    if (index + 1 <= imgs.length - 1) index++;
    else index = 0;
    return imgs[index];
  }

  // close lightbox
  function closeLightbox() {
    focusBody();

    const lightbox = document.getElementById("lightbox_overlay");
    if (lightbox) lightbox.remove();
  }

  // make all elements behind lightbox non-focusable
  function blurBody(overlay) {
    const all = document.querySelectorAll("*");
    for (const element of all) element.tabIndex = -1;
    document.body.classList.add("body_no_scroll");
  }

  // make all elements focusable again
  function focusBody() {
    const all = document.querySelectorAll("*");
    for (const element of all) element.removeAttribute("tabIndex");
    document.body.classList.remove("body_no_scroll");
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* regular <img> in document when hovered */
    img.lightbox_document_img:hover {
      cursor: pointer;
    }

    .body_no_scroll {
      overflow: hidden !important;
    }

    /* screen overlay */
    #lightbox_overlay {
      display: flex;
      flex-direction: column;
      position: fixed;
      left: 0;
      top: 0;
      right: 0;
      bottom: 0;
      background: rgba(0, 0, 0, 0.75);
      z-index: 3;
    }

    /* middle area containing lightbox image */
    #lightbox_image_container {
      flex-grow: 1;
      display: flex;
      justify-content: center;
      align-items: center;
      overflow: hidden;
      position: relative;
      padding: 20px;
    }

    /* bottom area containing caption */
    #lightbox_bottom_container {
      display: flex;
      justify-content: center;
      align-items: center;
      height: 100px;
      min-height: 100px;
      max-height: 100px;
      background: rgba(0, 0, 0, 0.5);
    }

    /* image number info text box */
    #lightbox_number_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      left: 2px;
      top: 0;
      z-index: 4;
    }

    /* zoom info text box */
    #lightbox_zoom_info {
      position: absolute;
      color: #ffffff;
      font-weight: 600;
      right: 2px;
      top: 0;
      z-index: 4;
    }

    /* copy of image caption */
    #lightbox_caption {
      box-sizing: border-box;
      display: inline-block;
      width: 100%;
      max-height: 100%;
      padding: 10px 0;
      text-align: center;
      overflow-y: auto;
      color: #ffffff;
    }

    /* navigation previous/next button */
    .lightbox_button {
      width: 100px;
      height: 100%;
      min-width: 100px;
      min-height: 100%;
      color: #ffffff;
    }

    /* navigation previous/next button when hovered */
    .lightbox_button:hover {
      background: none !important;
    }

    /* navigation button icon */
    .lightbox_button > svg {
      height: 25px;
    }

    /* figure auto-number */
    #lightbox_caption > span:first-of-type {
      font-weight: bold;
      margin-right: 5px;
    }

    /* lightbox image when hovered */
    #lightbox_img:hover {
      cursor: grab;
    }

    /* lightbox image when grabbed */
    #lightbox_img:active {
      cursor: grabbing;
    }
  }

  /* when on screen < 480px wide */
  @media only screen and (max-width: 480px) {
    /* make navigation buttons skinnier on small screens to make more room for caption text */
    .lightbox_button {
      width: 50px;
      min-width: 50px;
    }
  }

  /* always hide lightbox on print */
  @media only print {
    #lightbox_overlay {
      display: none;
    }
  }
</style>
<!-- 
  Link Highlight Plugin

  Makes it such that when a user hovers or focuses a link, other links that have
  the same target will be highlighted. It also makes it such that when clicking
  a link, the target of the link (eg reference, figure, table) is briefly
  highlighted.
-->

<script type="module">
  // whether to also highlight links that go to external urls
  const externalLinks = "false";
  // whether user must click off to unhighlight instead of just
  // un-hovering
  const clickUnhighlight = "false";
  // whether to also highlight links that are unique
  const highlightUnique = "true";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach mouse and focus listeners to link
      link.addEventListener("mouseenter", onLinkFocus);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("mouseleave", onLinkUnhover);
    }

    // attach click and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("hashchange", onHashChange);

    // run hash change on window load in case user has navigated
    // directly to hash
    onHashChange();
  }

  // when link is focused (tabbed to) or hovered
  function onLinkFocus() {
    highlight(this);
  }

  // when link is unhovered
  function onLinkUnhover() {
    if (clickUnhighlight !== "true") unhighlightAll();
  }

  // when the mouse is clicked anywhere in window
  function onClick(event) {
    unhighlightAll();
  }

  // when hash (eg manuscript.html#introduction) changes
  function onHashChange() {
    const target = getHashTarget();
    if (target) glowElement(target);
  }

  // start glow sequence on an element
  function glowElement(element) {
    const startGlow = function () {
      onGlowEnd();
      element.dataset.glow = "true";
      element.addEventListener("animationend", onGlowEnd);
    };
    const onGlowEnd = function () {
      element.removeAttribute("data-glow");
      element.removeEventListener("animationend", onGlowEnd);
    };
    startGlow();
  }

  // highlight link and all others with same target
  function highlight(link) {
    // force unhighlight all to start fresh
    unhighlightAll();

    // get links with same target
    if (!link) return;
    const sameLinks = getSameLinks(link);

    // if link unique and option is off, exit and don't highlight
    if (sameLinks.length <= 1 && highlightUnique !== "true") return;

    // highlight all same links, and "select" (special highlight) this
    // one
    for (const sameLink of sameLinks) {
      if (sameLink === link) sameLink.setAttribute("data-selected", "true");
      else sameLink.setAttribute("data-highlighted", "true");
    }
  }

  // unhighlight all links
  function unhighlightAll() {
    const links = getLinks();
    for (const link of links) {
      link.setAttribute("data-selected", "false");
      link.setAttribute("data-highlighted", "false");
    }
  }

  // get links with same target
  function getSameLinks(link) {
    const results = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        results.push(otherLink);
    }
    return results;
  }

  // get all links of types we wish to handle
  function getLinks() {
    let query = "a";
    if (externalLinks !== "true") query += '[href^="#"]';
    // exclude buttons, anchor links, toc links, etc
    query += ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    return document.querySelectorAll(query);
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<style>
  @media only screen {
    /* anything with data-highlighted attribute true */
    [data-highlighted="true"] {
      background: #ffeb3b;
    }

    /* anything with data-selected attribute true */
    [data-selected="true"] {
      background: #ff8a65 !important;
    }

    /* animation definition for glow */
    @keyframes highlight_glow {
      0% {
        background: none;
      }
      10% {
        background: #bbdefb;
      }
      100% {
        background: none;
      }
    }

    /* anything with data-glow attribute true */
    [data-glow="true"] {
      animation: highlight_glow 2s;
    }
  }
</style>
<!--
  Table of Contents Plugin

  Provides a "table of contents" (toc) panel on the side of the document that
  allows the user to conveniently navigate between sections of the document.
-->

<script type="module">
  // which types of elements to add links for, in "document.querySelector" format
  const typesQuery = "h1, h2, h3";
  // whether toc starts open. use 'true' or 'false', or 'auto' to
  // use 'true' behavior when screen wide enough and 'false' when not
  const startOpen = "false";
  // whether toc closes when clicking on toc link. use 'true' or
  // 'false', or 'auto' to use 'false' behavior when screen wide
  // enough and 'true' when not
  const clickClose = "auto";
  // if list item is more than this many characters, text will be
  // truncated
  const charLimit = "50";
  // whether or not to show bullets next to each toc item
  const bullets = "false";

  // start script
  function start() {
    // make toc panel and populate with entries (links to document
    // sections)
    const panel = makePanel();
    if (!panel) return;
    makeEntries(panel);
    // attach panel to document after making entries, so 'toc' heading
    // in panel isn't included in toc
    document.body.insertBefore(panel, document.body.firstChild);

    // initial panel state
    if (startOpen === "true" || (startOpen === "auto" && !isSmallScreen()))
      openPanel();
    else closePanel();

    // attach click, scroll, and hash change listeners to window
    window.addEventListener("click", onClick);
    window.addEventListener("scroll", onScroll);
    window.addEventListener("hashchange", onScroll);
    window.addEventListener("keyup", onKeyUp);
    onScroll();

    // add class to push document body down out of way of toc button
    document.body.classList.add("toc_body_nudge");
  }

  // determine if screen wide enough to fit toc panel
  function isSmallScreen() {
    // in default theme:
    // 816px = 8.5in = width of "page" (<body>) element
    // 260px = min width of toc panel (*2 for both sides of <body>)
    return window.innerWidth < 816 + 260 * 2;
  }

  // when mouse is clicked anywhere in window
  function onClick() {
    if (isSmallScreen()) closePanel();
  }

  // when window is scrolled or hash changed
  function onScroll() {
    highlightViewed();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    // close on esc
    if (event.key === "Escape") closePanel();
  }

  // find entry of currently viewed document section in toc and highlight
  function highlightViewed() {
    const firstId = getFirstInView(typesQuery);

    // get toc entries (links), unhighlight all, then highlight viewed
    const list = document.getElementById("toc_list");
    if (!firstId || !list) return;
    const links = list.querySelectorAll("a");
    for (const link of links) link.dataset.viewing = "false";
    const link = list.querySelector('a[href="#' + firstId + '"]');
    if (!link) return;
    link.dataset.viewing = "true";
  }

  // get first or previous toc listed element in top half of view
  function getFirstInView(query) {
    // get all elements matching query and with id
    const elements = document.querySelectorAll(query);
    const elementsWithIds = [];
    for (const element of elements) {
      if (element.id) elementsWithIds.push(element);
    }

    // get first or previous element in top half of view
    for (let i = 0; i < elementsWithIds.length; i++) {
      const element = elementsWithIds[i];
      const prevElement = elementsWithIds[Math.max(0, i - 1)];
      if (element.getBoundingClientRect().top >= 0) {
        if (element.getBoundingClientRect().top < window.innerHeight / 2)
          return element.id;
        else return prevElement.id;
      }
    }
  }

  // make panel
  function makePanel() {
    // create panel
    const panel = document.createElement("div");
    panel.id = "toc_panel";
    if (bullets === "true") panel.dataset.bullets = "true";

    // create header
    const header = document.createElement("div");
    header.id = "toc_header";

    // create toc button
    const button = document.createElement("button");
    button.id = "toc_button";
    button.innerHTML = document.querySelector(".icon_th_list").innerHTML;
    button.title = "Table of Contents";
    button.classList.add("icon_button");

    // create header text
    const text = document.createElement("h4");
    text.innerHTML = "Table of Contents";

    // create container for toc list
    const list = document.createElement("div");
    list.id = "toc_list";

    // attach click listeners
    panel.addEventListener("click", onPanelClick);
    header.addEventListener("click", onHeaderClick);
    button.addEventListener("click", onButtonClick);

    // attach elements
    header.appendChild(button);
    header.appendChild(text);
    panel.appendChild(header);
    panel.appendChild(list);

    return panel;
  }

  // create toc entries (links) to each element of the specified types
  function makeEntries(panel) {
    const elements = document.querySelectorAll(typesQuery);
    for (const element of elements) {
      // do not add link if element doesn't have assigned id
      if (!element.id) continue;

      // create link/list item
      const link = document.createElement("a");
      link.classList.add("toc_link");
      switch (element.tagName.toLowerCase()) {
        case "h1":
          link.dataset.level = "1";
          break;
        case "h2":
          link.dataset.level = "2";
          break;
        case "h3":
          link.dataset.level = "3";
          break;
        case "h4":
          link.dataset.level = "4";
          break;
      }
      link.title = element.innerText;
      let text = element.innerText;
      if (text.length > charLimit) text = text.slice(0, charLimit) + "...";
      link.innerHTML = text;
      link.href = "#" + element.id;
      link.addEventListener("click", onLinkClick);

      // attach link
      panel.querySelector("#toc_list").appendChild(link);
    }
  }

  // when panel is clicked
  function onPanelClick(event) {
    // stop click from propagating to window/document and closing panel
    event.stopPropagation();
  }

  // when header itself is clicked
  function onHeaderClick(event) {
    togglePanel();
  }

  // when button is clicked
  function onButtonClick(event) {
    togglePanel();
    // stop header underneath button from also being clicked
    event.stopPropagation();
  }

  // when link is clicked
  function onLinkClick(event) {
    if (clickClose === "true" || (clickClose === "auto" && isSmallScreen()))
      closePanel();
    else openPanel();
  }

  // open panel if closed, close if opened
  function togglePanel() {
    const panel = document.getElementById("toc_panel");
    if (!panel) return;

    if (panel.dataset.open === "true") closePanel();
    else openPanel();
  }

  // open panel
  function openPanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "true";
  }

  // close panel
  function closePanel() {
    const panel = document.getElementById("toc_panel");
    if (panel) panel.dataset.open = "false";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- th list icon -->

<template class="icon_th_list">
  <!-- modified from: https://fontawesome.com/icons/th-list -->
  <svg width="16" height="16" viewBox="0 0 512 512" tabindex="-1">
    <path
      fill="currentColor"
      d="M96 96c0 26.51-21.49 48-48 48S0 122.51 0 96s21.49-48 48-48 48 21.49 48 48zM48 208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm0 160c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm96-236h352c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h352c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H144c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* toc panel */
    #toc_panel {
      box-sizing: border-box;
      position: fixed;
      top: 0;
      left: 0;
      background: #ffffff;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      z-index: 2;
    }

    /* toc panel when closed */
    #toc_panel[data-open="false"] {
      min-width: 60px;
      width: 60px;
      height: 60px;
      border-right: solid 1px #bdbdbd;
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc panel when open */
    #toc_panel[data-open="true"] {
      min-width: 260px;
      max-width: 480px;
      /* keep panel edge consistent distance away from "page" edge */
      width: calc(((100vw - 8.5in) / 2) - 30px - 40px);
      bottom: 0;
      border-right: solid 1px #bdbdbd;
    }

    /* toc panel header */
    #toc_header {
      box-sizing: border-box;
      display: flex;
      flex-direction: row;
      align-items: center;
      height: 60px;
      margin: 0;
      padding: 20px;
    }

    /* toc panel header when hovered */
    #toc_header:hover {
      cursor: pointer;
    }

    /* toc panel header when panel open */
    #toc_panel[data-open="true"] > #toc_header {
      border-bottom: solid 1px #bdbdbd;
    }

    /* toc open/close header button */
    #toc_button {
      margin-right: 20px;
    }

    /* hide toc list and header text when closed */
    #toc_panel[data-open="false"] > #toc_header > *:not(#toc_button),
    #toc_panel[data-open="false"] > #toc_list {
      display: none;
    }

    /* toc list of entries */
    #toc_list {
      box-sizing: border-box;
      width: 100%;
      padding: 20px;
      position: absolute;
      top: calc(60px + 1px);
      bottom: 0;
      overflow: auto;
    }

    /* toc entry, link to section in document */
    .toc_link {
      display: block;
      padding: 5px;
      position: relative;
      font-weight: 600;
      text-decoration: none;
    }

    /* toc entry when hovered or when "viewed" */
    .toc_link:hover,
    .toc_link[data-viewing="true"] {
      background: #f5f5f5;
    }

    /* toc entry, level 1 indentation */
    .toc_link[data-level="1"] {
      margin-left: 0;
    }

    /* toc entry, level 2 indentation */
    .toc_link[data-level="2"] {
      margin-left: 20px;
    }

    /* toc entry, level 3 indentation */
    .toc_link[data-level="3"] {
      margin-left: 40px;
    }

    /* toc entry, level 4 indentation */
    .toc_link[data-level="4"] {
      margin-left: 60px;
    }

    /* toc entry bullets */
    #toc_panel[data-bullets="true"] .toc_link[data-level]:before {
      position: absolute;
      left: -15px;
      top: -1px;
      font-size: 1.5em;
    }

    /* toc entry, level 2 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="2"]:before {
      content: "\2022";
    }

    /* toc entry, level 3 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="3"]:before {
      content: "\25AB";
    }

    /* toc entry, level 4 bullet */
    #toc_panel[data-bullets="true"] .toc_link[data-level="4"]:before {
      content: "-";
    }
  }

  /* when on screen < 8.5in wide */
  @media only screen and (max-width: 8.5in) {
    /* push <body> ("page") element down to make room for toc icon */
    .toc_body_nudge {
      padding-top: 60px;
    }

    /* toc icon when panel closed and not hovered */
    #toc_panel[data-open="false"]:not(:hover) {
      background: rgba(255, 255, 255, 0.75);
    }
  }

  /* always hide toc panel on print */
  @media only print {
    #toc_panel {
      display: none;
    }
  }
</style>
<!-- 
  Tooltips Plugin

  Makes it such that when the user hovers or focuses a link to a citation or
  figure, a tooltip appears with a preview of the reference content, along with
  arrows to navigate between instances of the same reference in the document.
-->

<script type="module">
  // whether user must click off to close tooltip instead of just un-hovering
  const clickClose = "false";
  // delay (in ms) between opening and closing tooltip
  const delay = "100";

  // start script
  function start() {
    const links = getLinks();
    for (const link of links) {
      // attach hover and focus listeners to link
      link.addEventListener("mouseover", onLinkHover);
      link.addEventListener("mouseleave", onLinkUnhover);
      link.addEventListener("focus", onLinkFocus);
      link.addEventListener("touchend", onLinkTouch);
    }

    // attach mouse, key, and resize listeners to window
    window.addEventListener("mousedown", onClick);
    window.addEventListener("touchstart", onClick);
    window.addEventListener("keyup", onKeyUp);
    window.addEventListener("resize", onResize);
  }

  // when link is hovered
  function onLinkHover() {
    // function to open tooltip
    const delayOpenTooltip = function () {
      openTooltip(this);
    }.bind(this);

    // run open function after delay
    this.openTooltipTimer = window.setTimeout(delayOpenTooltip, delay);
  }

  // when mouse leaves link
  function onLinkUnhover() {
    // cancel opening tooltip
    window.clearTimeout(this.openTooltipTimer);

    // don't close on unhover if option specifies
    if (clickClose === "true") return;

    // function to close tooltip
    const delayCloseTooltip = function () {
      // if tooltip open and if mouse isn't over tooltip, close
      const tooltip = document.getElementById("tooltip");
      if (tooltip && !tooltip.matches(":hover")) closeTooltip();
    };

    // run close function after delay
    this.closeTooltipTimer = window.setTimeout(delayCloseTooltip, delay);
  }

  // when link is focused (tabbed to)
  function onLinkFocus(event) {
    openTooltip(this);
  }

  // when link is touched on touch screen
  function onLinkTouch(event) {
    // attempt to force hover state on first tap always, and trigger
    // regular link click (and navigation) on second tap
    if (event.target === document.activeElement) event.target.click();
    else {
      document.activeElement.blur();
      event.target.focus();
    }
    if (event.cancelable) event.preventDefault();
    event.stopPropagation();
    return false;
  }

  // when mouse is clicked anywhere in window
  function onClick(event) {
    closeTooltip();
  }

  // when key pressed
  function onKeyUp(event) {
    if (!event || !event.key) return;

    switch (event.key) {
      // trigger click of prev button
      case "ArrowLeft":
        const prevButton = document.getElementById("tooltip_prev_button");
        if (prevButton) prevButton.click();
        break;
      // trigger click of next button
      case "ArrowRight":
        const nextButton = document.getElementById("tooltip_next_button");
        if (nextButton) nextButton.click();
        break;
      // close on esc
      case "Escape":
        closeTooltip();
        break;
    }
  }

  // when window is resized or zoomed
  function onResize() {
    closeTooltip();
  }

  // get all links of types we wish to handle
  function getLinks() {
    const queries = [];
    // exclude buttons, anchor links, toc links, etc
    const exclude =
      ":not(.button):not(.icon_button):not(.anchor):not(.toc_link)";
    queries.push('a[href^="#ref-"]' + exclude); // citation links
    queries.push('a[href^="#fig:"]' + exclude); // figure links
    const query = queries.join(", ");
    return document.querySelectorAll(query);
  }

  // get links with same target, get index of link in set, get total
  // same links
  function getSameLinks(link) {
    const sameLinks = [];
    const links = getLinks();
    for (const otherLink of links) {
      if (otherLink.getAttribute("href") === link.getAttribute("href"))
        sameLinks.push(otherLink);
    }

    return {
      elements: sameLinks,
      index: sameLinks.indexOf(link),
      total: sameLinks.length,
    };
  }

  // open tooltip
  function openTooltip(link) {
    // delete tooltip if it exists, start fresh
    closeTooltip();

    // make tooltip element
    const tooltip = makeTooltip(link);

    // if source couldn't be found and tooltip not made, exit
    if (!tooltip) return;

    // make navbar elements
    const navBar = makeNavBar(link);
    if (navBar) tooltip.firstElementChild.appendChild(navBar);

    // attach tooltip to page
    document.body.appendChild(tooltip);

    // position tooltip
    const position = function () {
      positionTooltip(link);
    };
    position();

    // if tooltip contains images, position again after they've loaded
    const imgs = tooltip.querySelectorAll("img");
    for (const img of imgs) img.addEventListener("load", position);
  }

  // close (delete) tooltip
  function closeTooltip() {
    const tooltip = document.getElementById("tooltip");
    if (tooltip) tooltip.remove();
  }

  // make tooltip
  function makeTooltip(link) {
    // get target element that link points to
    const source = getSource(link);

    // if source can't be found, exit
    if (!source) return;

    // create new tooltip
    const tooltip = document.createElement("div");
    tooltip.id = "tooltip";
    const tooltipContent = document.createElement("div");
    tooltipContent.id = "tooltip_content";
    tooltip.appendChild(tooltipContent);

    // make copy of source node and put in tooltip
    const sourceCopy = makeCopy(source);
    tooltipContent.appendChild(sourceCopy);

    // attach mouse event listeners
    tooltip.addEventListener("click", onTooltipClick);
    tooltip.addEventListener("mousedown", onTooltipClick);
    tooltip.addEventListener("touchstart", onTooltipClick);
    tooltip.addEventListener("mouseleave", onTooltipUnhover);

    // (for interaction with lightbox plugin)
    // transfer click on tooltip copied img to original img
    const sourceImg = source.querySelector("img");
    const sourceCopyImg = sourceCopy.querySelector("img");
    if (sourceImg && sourceCopyImg) {
      const clickImg = function () {
        sourceImg.click();
        closeTooltip();
      };
      sourceCopyImg.addEventListener("click", clickImg);
    }

    return tooltip;
  }

  // make carbon copy of html dom element
  function makeCopy(source) {
    const sourceCopy = source.cloneNode(true);

    // delete elements marked with ignore (eg anchor and jump buttons)
    const deleteFromCopy = sourceCopy.querySelectorAll('[data-ignore="true"]');
    for (const element of deleteFromCopy) element.remove();

    // delete certain element attributes
    const attributes = [
      "id",
      "data-collapsed",
      "data-selected",
      "data-highlighted",
      "data-glow",
      "class",
    ];
    for (const attribute of attributes) {
      sourceCopy.removeAttribute(attribute);
      const elements = sourceCopy.querySelectorAll("[" + attribute + "]");
      for (const element of elements) element.removeAttribute(attribute);
    }

    return sourceCopy;
  }

  // when tooltip is clicked
  function onTooltipClick(event) {
    // when user clicks on tooltip, stop click from transferring
    // outside of tooltip (eg, click off to close tooltip, or eg click
    // off to unhighlight same refs)
    event.stopPropagation();
  }

  // when tooltip is unhovered
  function onTooltipUnhover(event) {
    if (clickClose === "true") return;

    // make sure new mouse/touch/focus no longer over tooltip or any
    // element within it
    const tooltip = document.getElementById("tooltip");
    if (!tooltip) return;
    if (this.contains(event.relatedTarget)) return;

    closeTooltip();
  }

  // make nav bar to go betwen prev/next instances of same reference
  function makeNavBar(link) {
    // find other links to the same source
    const sameLinks = getSameLinks(link);

    // don't show nav bar when singular reference
    if (sameLinks.total <= 1) return;

    // find prev/next links with same target
    const prevLink = getPrevLink(link, sameLinks);
    const nextLink = getNextLink(link, sameLinks);

    // create nav bar
    const navBar = document.createElement("div");
    navBar.id = "tooltip_nav_bar";
    const text = sameLinks.index + 1 + " of " + sameLinks.total;

    // create nav bar prev/next buttons
    const prevButton = document.createElement("button");
    const nextButton = document.createElement("button");
    prevButton.id = "tooltip_prev_button";
    nextButton.id = "tooltip_next_button";
    prevButton.title =
      "Jump to the previous occurence of this item in the document [←]";
    nextButton.title =
      "Jump to the next occurence of this item in the document [→]";
    prevButton.classList.add("icon_button");
    nextButton.classList.add("icon_button");
    prevButton.innerHTML = document.querySelector(".icon_caret_left").innerHTML;
    nextButton.innerHTML =
      document.querySelector(".icon_caret_right").innerHTML;
    navBar.appendChild(prevButton);
    navBar.appendChild(document.createTextNode(text));
    navBar.appendChild(nextButton);

    // attach click listeners to buttons
    prevButton.addEventListener("click", function () {
      onPrevNextClick(link, prevLink);
    });
    nextButton.addEventListener("click", function () {
      onPrevNextClick(link, nextLink);
    });

    return navBar;
  }

  // get previous link with same target
  function getPrevLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if < 1
    let index;
    if (sameLinks.index - 1 >= 0) index = sameLinks.index - 1;
    else index = sameLinks.total - 1;
    return sameLinks.elements[index];
  }

  // get next link with same target
  function getNextLink(link, sameLinks) {
    if (!sameLinks) sameLinks = getSameLinks(link);
    // wrap index to other side if > total
    let index;
    if (sameLinks.index + 1 <= sameLinks.total - 1) index = sameLinks.index + 1;
    else index = 0;
    return sameLinks.elements[index];
  }

  // get element that is target of link or url hash
  function getSource(link) {
    const hash = link ? link.hash : window.location.hash;
    const id = hash.slice(1);
    let target = document.querySelector('[id="' + id + '"]');
    if (!target) return;

    // if ref or figure, modify target to get expected element
    if (id.indexOf("ref-") === 0) target = target.querySelector(":nth-child(2)");
    else if (id.indexOf("fig:") === 0) target = target.querySelector("figure");

    return target;
  }

  // when prev/next arrow button is clicked
  function onPrevNextClick(link, prevNextLink) {
    if (link && prevNextLink)
      goToElement(prevNextLink, window.innerHeight * 0.5);
  }

  // scroll to and focus element
  function goToElement(element, offset) {
    // expand accordion section if collapsed
    expandElement(element);
    const y =
      getRectInView(element).top -
      getRectInView(document.documentElement).top -
      (offset || 0);
    // trigger any function listening for "onscroll" event
    window.dispatchEvent(new Event("scroll"));
    window.scrollTo(0, y);
    document.activeElement.blur();
    element.focus();
  }

  // determine position to place tooltip based on link position in
  // viewport and tooltip size
  function positionTooltip(link, left, top) {
    const tooltipElement = document.getElementById("tooltip");
    if (!tooltipElement) return;

    // get convenient vars for position/dimensions of
    // link/tooltip/page/view
    link = getRectInPage(link);
    const tooltip = getRectInPage(tooltipElement);
    const view = getRectInPage();

    // horizontal positioning
    if (left)
      // use explicit value
      left = left;
    else if (link.left + tooltip.width < view.right)
      // fit tooltip to right of link
      left = link.left;
    else if (link.right - tooltip.width > view.left)
      // fit tooltip to left of link
      left = link.right - tooltip.width;
    // center tooltip in view
    else left = (view.right - view.left) / 2 - tooltip.width / 2;

    // vertical positioning
    if (top)
      // use explicit value
      top = top;
    else if (link.top - tooltip.height > view.top)
      // fit tooltip above link
      top = link.top - tooltip.height;
    else if (link.bottom + tooltip.height < view.bottom)
      // fit tooltip below link
      top = link.bottom;
    else {
      // center tooltip in view
      top = view.top + view.height / 2 - tooltip.height / 2;
      // nudge off of link to left/right if possible
      if (link.right + tooltip.width < view.right) left = link.right;
      else if (link.left - tooltip.width > view.left)
        left = link.left - tooltip.width;
    }

    tooltipElement.style.left = left + "px";
    tooltipElement.style.top = top + "px";
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- caret left icon -->

<template class="icon_caret_left">
  <!-- modified from: https://fontawesome.com/icons/caret-left -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
    ></path>
  </svg>
</template>

<!-- caret right icon -->

<template class="icon_caret_right">
  <!-- modified from: https://fontawesome.com/icons/caret-right -->
  <svg width="16" height="16" viewBox="0 0 192 512">
    <path
      fill="currentColor"
      d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
    ></path>
  </svg>
</template>

<style>
  @media only screen {
    /* tooltip container */
    #tooltip {
      position: absolute;
      width: 50%;
      min-width: 240px;
      max-width: 75%;
      z-index: 1;
    }

    /* tooltip content */
    #tooltip_content {
      margin-bottom: 5px;
      padding: 20px;
      border-radius: 5px;
      border: solid 1px #bdbdbd;
      box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
      background: #ffffff;
      overflow-wrap: break-word;
    }

    /* tooltip copy of paragraphs and figures */
    #tooltip_content > p,
    #tooltip_content > figure {
      margin: 0;
      max-height: 320px;
      overflow-y: auto;
    }

    /* tooltip copy of <img> */
    #tooltip_content > figure > img,
    #tooltip_content > figure > svg {
      max-height: 260px;
    }

    /* navigation bar */
    #tooltip_nav_bar {
      margin-top: 10px;
      text-align: center;
    }

    /* navigation bar previous/next buton */
    #tooltip_nav_bar > .icon_button {
      position: relative;
      top: 3px;
    }

    /* navigation bar previous button */
    #tooltip_nav_bar > .icon_button:first-of-type {
      margin-right: 5px;
    }

    /* navigation bar next button */
    #tooltip_nav_bar > .icon_button:last-of-type {
      margin-left: 5px;
    }
  }

  /* always hide tooltip on print */
  @media only print {
    #tooltip {
      display: none;
    }
  }
</style>
<!--
  Analytics Plugin (third-party) 
  
  Copy and paste code from Google Analytics or similar service here.
-->
<!-- 
  Annotations Plugin

  Allows public annotation of the  manuscript. See https://web.hypothes.is/.
-->

<script type="module">
  // configuration
  window.hypothesisConfig = function () {
    return {
      branding: {
        accentColor: "#2196f3",
        appBackgroundColor: "#f8f8f8",
        ctaBackgroundColor: "#f8f8f8",
        ctaTextColor: "#000000",
        selectionFontFamily: "Open Sans, Helvetica, sans serif",
        annotationFontFamily: "Open Sans, Helvetica, sans serif",
      },
    };
  };

  // hypothesis client script
  const embed = "https://hypothes.is/embed.js";
  // hypothesis annotation count query url
  const query = "https://api.hypothes.is/api/search?limit=0&url=";

  // start script
  function start() {
    const button = makeButton();
    document.body.insertBefore(button, document.body.firstChild);
    insertCount(button);
  }

  // make button
  function makeButton() {
    // create button
    const button = document.createElement("button");
    button.id = "hypothesis_button";
    button.innerHTML = document.querySelector(".icon_hypothesis").innerHTML;
    button.title = "Hypothesis annotations";
    button.classList.add("icon_button");

    function onClick(event) {
      onButtonClick(event, button);
    }

    // attach click listeners
    button.addEventListener("click", onClick);

    return button;
  }

  // insert annotations count
  async function insertCount(button) {
    // get annotation count from Hypothesis based on url
    let count = "-";
    try {
      const canonical = document.querySelector('link[rel="canonical"]');
      const location = window.location;
      const url = encodeURIComponent((canonical || location).href);
      const response = await fetch(query + url);
      const json = await response.json();
      count = json.total || "-";
    } catch (error) {
      console.log(error);
    }

    // put count into button
    const counter = document.createElement("span");
    counter.id = "hypothesis_count";
    counter.innerHTML = count;
    button.title = "View " + count + " Hypothesis annotations";
    button.append(counter);
  }

  // when button is clicked
  function onButtonClick(event, button) {
    const script = document.createElement("script");
    script.src = embed;
    document.body.append(script);
    button.remove();
  }

  // start script when document is finished loading
  window.addEventListener("load", start);
</script>

<!-- hypothesis icon -->

<template class="icon_hypothesis">
  <!-- modified from: https://simpleicons.org/icons/hypothesis.svg / https://git.io/Jf1VB -->
  <svg width="16" height="16" viewBox="0 0 24 24" tabindex="-1">
    <path
      fill="currentColor"
      d="M3.43 0C2.5 0 1.72 .768 1.72 1.72V18.86C1.72 19.8 2.5 20.57 3.43 20.57H9.38L12 24L14.62 20.57H20.57C21.5 20.57 22.29 19.8 22.29 18.86V1.72C22.29 .77 21.5 0 20.57 0H3.43M5.14 3.43H7.72V9.43S8.58 7.72 10.28 7.72C12 7.72 13.74 8.57 13.74 11.24V17.14H11.16V12C11.16 10.61 10.28 10.07 9.43 10.29C8.57 10.5 7.72 11.41 7.72 13.29V17.14H5.14V3.43M18 13.72C18.95 13.72 19.72 14.5 19.72 15.42A1.71 1.71 0 0 1 18 17.13A1.71 1.71 0 0 1 16.29 15.42C16.29 14.5 17.05 13.71 18 13.71Z"
      tabindex="-1"
    ></path>
  </svg>
</template>

<style>
  /* hypothesis activation button */
  #hypothesis_button {
    box-sizing: border-box;
    position: fixed;
    top: 0;
    right: 0;
    width: 60px;
    height: 60px;
    background: #ffffff;
    border-radius: 0;
    border-left: solid 1px #bdbdbd;
    border-bottom: solid 1px #bdbdbd;
    box-shadow: 0 0 20px rgba(0, 0, 0, 0.05);
    z-index: 2;
  }

  /* hypothesis button svg */
  #hypothesis_button > svg {
    position: relative;
    top: -4px;
  }

  /* hypothesis annotation count */
  #hypothesis_count {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 5px;
  }

  /* side panel */
  .annotator-frame {
    width: 280px !important;
  }

  /* match highlight color to rest of theme */
  .annotator-highlights-always-on .annotator-hl {
    background-color: #ffeb3b !important;
  }

  /* match focused color to rest of theme */
  .annotator-hl.annotator-hl-focused {
    background-color: #ff8a65 !important;
  }

  /* match bucket bar color to rest of theme */
  .annotator-bucket-bar {
    background: #f5f5f5 !important;
  }

  /* always hide button, toolbar, and tooltip on print */
  @media only print {
    #hypothesis_button {
      display: none;
    }

    .annotator-frame {
      display: none !important;
    }

    hypothesis-adder {
      display: none !important;
    }
  }
</style>
<!-- 
  Mathjax Plugin (third-party) 

  Allows the proper rendering of math/equations written in LaTeX.
  See https://www.mathjax.org/.
-->

<script type="text/x-mathjax-config">
  // configuration
  MathJax.Hub.Config({
    "CommonHTML": { linebreaks: { automatic: true } },
    "HTML-CSS": { linebreaks: { automatic: true } },
    "SVG": { linebreaks: { automatic: true } },
    "fast-preview": { disabled: true }
  });
</script>

<script
  src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"
  integrity="sha512-0doc9hKxR3PYwso42RD1p5ySZpzzuDiOwMrdCEh2WdJZCjcmFKc/wEnL+z8fBQrnHoiNWbo+3fiGkOYXBdQp4A=="
  crossorigin="anonymous"
></script>

<style>
  /* mathjax containers */
  .math.display > span:not(.MathJax_Preview) {
    /* turn inline element (no dimensions) into block (allows fixed width and thus scrolling) */
    display: flex !important;
    overflow-x: auto !important;
    overflow-y: hidden !important;
    justify-content: center;
    align-items: center;
    margin: 0 !important;
  }

  /* right click menu */
  .MathJax_Menu {
    border-radius: 5px !important;
    border: solid 1px #bdbdbd !important;
    box-shadow: none !important;
  }

  /* equation auto-number */
  span[id^="eq:"] > span.math.display + span {
    font-weight: 600;
  }

  /* equation */
  span[id^="eq:"] > span.math.display > span {
    /* nudge to make room for equation auto-number and anchor */
    margin-right: 60px !important;
  }
</style>
</body>
</html>
